diff -up ./Eigen/src/StlSupport/details.h.ptrdiff ./Eigen/src/StlSupport/details.h
--- ./Eigen/src/StlSupport/details.h.ptrdiff	2011-04-17 23:07:20.116801391 -0400
+++ ./Eigen/src/StlSupport/details.h	2011-04-17 23:07:49.699857645 -0400
@@ -30,6 +30,8 @@
   #define EIGEN_ALIGNED_ALLOCATOR Eigen::aligned_allocator
 #endif
 
+#include <cstddef> // for ptrdiff_t
+
 namespace Eigen {
 
   // This one is needed to prevent reimplementing the whole std::vector.
