Index: modules/core/tools/container_templates/set.hpp
===================================================================
--- modules/core/tools/container_templates/set.hpp	(revision 1087)
+++ modules/core/tools/container_templates/set.hpp	(working copy)
@@ -11,9 +11,9 @@
 #ifndef IMPCORE_GROUPNAME_CONTAINER_SET_H
 #define IMPCORE_GROUPNAME_CONTAINER_SET_H
 
-#include "core_exports.h"
-#include "GroupnameContainer.h"
+#include "config.h"
 #include "internal/core_version_info.h"
+#include <IMP/GroupnameContainer.h>
 
 IMPCORE_BEGIN_NAMESPACE
 
Index: modules/core/tools/container_templates/restraint.cc
===================================================================
--- modules/core/tools/container_templates/restraint.cc	(revision 1087)
+++ modules/core/tools/container_templates/restraint.cc	(working copy)
@@ -10,7 +10,7 @@
  */
 
 #include "IMP/core/GroupnamesRestraint.h"
-#include "IMP/core/internal/container_helpers.h"
+#include <IMP/internal/container_helpers.h>
 #include "IMP/core/ListGroupnameContainer.h"
 
 #include <IMP/GroupnameScore.h>
@@ -50,7 +50,7 @@
   for (GroupnameContainer::ClassnameIterator
          it= pc_->classnames_begin();
        it != pc_->classnames_end(); ++it) {
-    score += internal::ContainerTraits<Classname>
+    score += IMP::internal::ContainerTraits<Classname>
       ::evaluate(ss_, *it, accum);
   }
 
@@ -60,12 +60,12 @@
 
 ParticlesList GroupnamesRestraint::get_interacting_particles() const
 {
-  if (!internal::ContainerTraits<Classname>::is_singleton) {
+  if (!IMP::internal::ContainerTraits<Classname>::is_singleton) {
     ParticlesList ret;
     for (GroupnameContainer::ClassnameIterator it
            = pc_->classnames_begin();
          it != pc_->classnames_end(); ++it) {
-      ret.push_back(internal::ContainerTraits<Classname>::create_set(*it));
+      ret.push_back(IMP::internal::ContainerTraits<Classname>::create_set(*it));
     }
     return ret;
   } else {
Index: modules/core/tools/container_templates/swig.i
===================================================================
--- modules/core/tools/container_templates/swig.i	(revision 1087)
+++ modules/core/tools/container_templates/swig.i	(working copy)
@@ -8,11 +8,11 @@
     IMP_OWN_FIRST_SECOND_THIRD_CONSTRUCTOR(GroupnamesScoreState)
     IMP_OWN_FIRST_SECOND_CONSTRUCTOR(MinimumGroupnameScoreRestraint)
     IMP_OWN_FIRST_SECOND_CONSTRUCTOR(MaximumGroupnameScoreRestraint)
-    IMPCORE_CONTAINER_SWIG(FilteredListGroupnameContainer, GroupnameContainer, groupname_container)
+    IMP_CONTAINER_SWIG(FilteredListGroupnameContainer, GroupnameContainer, groupname_container)
     IMP_SET_OBJECT(GroupnamesRestraint, set_groupname_container)
     IMP_SET_OBJECT(GroupnamesScoreState, set_before_evaluate_modifier)
     IMP_SET_OBJECT(GroupnamesScoreState, set_after_evaluate_modifier)
-    IMPCORE_CONTAINER_SWIG(GroupnameContainerSet, GroupnameContainer, groupname_container)
+    IMP_CONTAINER_SWIG(GroupnameContainerSet, GroupnameContainer, groupname_container)
 
     }
 }
Index: modules/core/tools/container_templates/restraint.hpp
===================================================================
--- modules/core/tools/container_templates/restraint.hpp	(revision 1087)
+++ modules/core/tools/container_templates/restraint.hpp	(working copy)
@@ -12,13 +12,13 @@
 #ifndef IMPCORE_GROUPNAMES_RESTRAINT_H
 #define IMPCORE_GROUPNAMES_RESTRAINT_H
 
-#include "core_exports.h"
+#include "config.h"
 #include "internal/core_version_info.h"
-#include "GroupnameContainer.h"
 
 #include <IMP/Restraint.h>
 #include <IMP/Pointer.h>
 #include <IMP/GroupnameScore.h>
+#include <IMP/GroupnameContainer.h>
 
 #include <iostream>
 
Index: modules/core/tools/container_templates/list.hpp
===================================================================
--- modules/core/tools/container_templates/list.hpp	(revision 1087)
+++ modules/core/tools/container_templates/list.hpp	(working copy)
@@ -10,9 +10,9 @@
 #ifndef IMPCORE_LIST_GROUPNAME_CONTAINER_H
 #define IMPCORE_LIST_GROUPNAME_CONTAINER_H
 
-#include "core_exports.h"
-#include "GroupnameContainer.h"
+#include "config.h"
 #include "internal/core_version_info.h"
+#include <IMP/GroupnameContainer.h>
 
 IMPCORE_BEGIN_NAMESPACE
 
Index: modules/core/tools/container_templates/state.cc
===================================================================
--- modules/core/tools/container_templates/state.cc	(revision 1087)
+++ modules/core/tools/container_templates/state.cc	(working copy)
@@ -11,7 +11,6 @@
  */
 
 #include "IMP/core/GroupnamesScoreState.h"
-#include "IMP/core/internal/container_helpers.h"
 
 IMPCORE_BEGIN_NAMESPACE
 
Index: modules/core/tools/container_templates/container.cc
===================================================================
--- modules/core/tools/container_templates/container.cc	(revision 1087)
+++ modules/core/tools/container_templates/container.cc	(working copy)
@@ -8,10 +8,10 @@
  *
  */
 
-#include "IMP/core/GroupnameContainer.h"
+#include "IMP/GroupnameContainer.h"
 
 
-IMPCORE_BEGIN_NAMESPACE
+IMP_BEGIN_NAMESPACE
 
 GroupnameContainer::GroupnameContainer(){}
 
@@ -22,4 +22,4 @@
 }
 
 
-IMPCORE_END_NAMESPACE
+IMP_END_NAMESPACE
Index: modules/core/tools/container_templates/state.hpp
===================================================================
--- modules/core/tools/container_templates/state.hpp	(revision 1087)
+++ modules/core/tools/container_templates/state.hpp	(working copy)
@@ -12,19 +12,19 @@
 #ifndef IMPCORE_GROUPNAMES_SCORE_STATE_H
 #define IMPCORE_GROUPNAMES_SCORE_STATE_H
 
-#include "core_exports.h"
-#include "GroupnameContainer.h"
-#include "GroupnameModifier.h"
+#include "config.h"
 #include "internal/core_version_info.h"
-
+#include <IMP/GroupnameContainer.h>
+#include <IMP/GroupnameModifier.h>
 #include <IMP/ScoreState.h>
 
-IMPCORE_BEGIN_NAMESPACE
-
+IMP_BEGIN_NAMESPACE
 // for swig
 class GroupnameContainer;
 class GroupnameModifier;
+IMP_END_NAMESPACE
 
+IMPCORE_BEGIN_NAMESPACE
 //! Apply a GroupnameFunction to a GroupnameContainer to maintain an invariant
 /** \ingroup restraint
     An example showing a how to use such a score state to maintain a cover
Index: modules/core/tools/container_templates/min.cc
===================================================================
--- modules/core/tools/container_templates/min.cc	(revision 1087)
+++ modules/core/tools/container_templates/min.cc	(working copy)
@@ -11,7 +11,7 @@
 
 #include "IMP/core/MinimumGroupnameScoreRestraint.h"
 #include "IMP/core/internal/MinimalSet.h"
-#include "IMP/core/internal/container_helpers.h"
+#include <IMP/internal/container_helpers.h>
 
 
 IMPCORE_BEGIN_NAMESPACE
@@ -31,7 +31,7 @@
             << std::distance(b,e) << std::endl);
     MS bestn(n);
     for (It it= b; it != e; ++it) {
-      float score= internal::ContainerTraits<Classname>
+      float score= IMP::internal::ContainerTraits<Classname>
                            ::evaluate(f, *it, NULL);
 
       if (bestn.can_insert(score)) {
@@ -53,7 +53,7 @@
   float score=0;
   for (unsigned int i=0; i< bestn.size(); ++i) {
     if (da) {
-      internal::ContainerTraits<Classname>::evaluate(f_.get(),
+      IMP::internal::ContainerTraits<Classname>::evaluate(f_.get(),
                                                      *bestn[i].second, da);
     }
     score+= bestn[i].first;
@@ -77,7 +77,7 @@
                              c_->classnames_end(), f_.get(), n_);
   ParticlesList ret;
   for (unsigned int i=0; i< bestn.size(); ++i) {
-    Particles c= internal::ContainerTraits<Classname>
+    Particles c= IMP::internal::ContainerTraits<Classname>
                  ::create_set(*bestn[i].second);
     ret.push_back(c);
   }
Index: modules/core/tools/container_templates/container.hpp
===================================================================
--- modules/core/tools/container_templates/container.hpp	(revision 1087)
+++ modules/core/tools/container_templates/container.hpp	(working copy)
@@ -10,20 +10,20 @@
 #ifndef IMPCORE_GROUPNAME_CONTAINER_H
 #define IMPCORE_GROUPNAME_CONTAINER_H
 
-#include "core_exports.h"
+#include "config.h"
 #include "internal/IndexingIterator.h"
 #include "internal/container_helpers.h"
-#include <IMP/Particle.h>
-#include <IMP/base_types.h>
-#include <IMP/Pointer.h>
+#include "Particle.h"
+#include "base_types.h"
+#include "Pointer.h"
 
-IMPCORE_BEGIN_NAMESPACE
+IMP_BEGIN_NAMESPACE
 
 //! A shared container for classnames
 /** Stores a searchable shared collection of classnames.
     \ingroup restraints
  */
-class IMPCOREEXPORT GroupnameContainer : public RefCountedObject
+class IMPEXPORT GroupnameContainer : public RefCountedObject
 {
   struct Accessor {
     typedef Accessor This;
@@ -89,7 +89,7 @@
 //! The index to use when this container is store in another object
 typedef Index<GroupnameContainer> GroupnameContainerIndex;
 
-IMPCORE_END_NAMESPACE
+IMP_END_NAMESPACE
 
 #define IMP_GROUPNAME_CONTAINER(version_info)                      \
   bool get_contains_classname(Value p) const;                      \
Index: modules/core/tools/container_templates/max.cc
===================================================================
--- modules/core/tools/container_templates/max.cc	(revision 1087)
+++ modules/core/tools/container_templates/max.cc	(working copy)
@@ -11,7 +11,7 @@
 
 #include "IMP/core/MaximumGroupnameScoreRestraint.h"
 #include "IMP/core/internal/MinimalSet.h"
-#include "IMP/core/internal/container_helpers.h"
+#include <IMP/internal/container_helpers.h>
 
 
 IMPCORE_BEGIN_NAMESPACE
@@ -32,7 +32,7 @@
     MS bestn(n);
     for (It it= b; it != e; ++it) {
       float score
-         = internal::ContainerTraits<Classname>::evaluate(f, *it, NULL);
+      = IMP::internal::ContainerTraits<Classname>::evaluate(f, *it, NULL);
 
       if (bestn.can_insert(score)) {
         bestn.insert(score, it);
@@ -57,7 +57,7 @@
   float score=0;
   for (unsigned int i=0; i< bestn.size(); ++i) {
     if (da) {
-      internal::ContainerTraits<Classname>::evaluate(f_.get(),
+      IMP::internal::ContainerTraits<Classname>::evaluate(f_.get(),
                                                      *bestn[i].second, da);
     }
     score+= bestn[i].first;
@@ -81,7 +81,7 @@
                              c_->classnames_end(), f_.get(), n_);
   ParticlesList ret;
   for (unsigned int i=0; i< bestn.size(); ++i) {
-    Particles c= internal::ContainerTraits<Classname>
+    Particles c= IMP::internal::ContainerTraits<Classname>
                          ::create_set(*bestn[i].second);
     ret.push_back(c);
   }
Index: modules/core/tools/container_templates/min.hpp
===================================================================
--- modules/core/tools/container_templates/min.hpp	(revision 1087)
+++ modules/core/tools/container_templates/min.hpp	(working copy)
@@ -11,11 +11,11 @@
 #ifndef IMPCORE_MINIMUM_GROUPNAME_SCORE_RESTRAINT_H
 #define IMPCORE_MINIMUM_GROUPNAME_SCORE_RESTRAINT_H
 
-#include "core_exports.h"
+#include "config.h"
 #include "internal/core_version_info.h"
-#include "GroupnameContainer.h"
 #include <IMP/Restraint.h>
 #include <IMP/GroupnameScore.h>
+#include <IMP/GroupnameContainer.h>
 
 IMPCORE_BEGIN_NAMESPACE
 
Index: modules/core/tools/container_templates/function.cc
===================================================================
--- modules/core/tools/container_templates/function.cc	(revision 1087)
+++ modules/core/tools/container_templates/function.cc	(working copy)
@@ -8,13 +8,13 @@
  *
  */
 
-#include "IMP/core/GroupnameModifier.h"
+#include "IMP/GroupnameModifier.h"
 
-IMPCORE_BEGIN_NAMESPACE
+IMP_BEGIN_NAMESPACE
 
 GroupnameModifier::GroupnameModifier(){
 }
 
 GroupnameModifier::~GroupnameModifier(){}
 
-IMPCORE_END_NAMESPACE
+IMP_END_NAMESPACE
Index: modules/core/tools/container_templates/max.hpp
===================================================================
--- modules/core/tools/container_templates/max.hpp	(revision 1087)
+++ modules/core/tools/container_templates/max.hpp	(working copy)
@@ -11,12 +11,13 @@
 #ifndef IMPCORE_MAXIMUM_GROUPNAME_SCORE_RESTRAINT_H
 #define IMPCORE_MAXIMUM_GROUPNAME_SCORE_RESTRAINT_H
 
-#include "core_exports.h"
+#include "config.h"
 #include "internal/core_version_info.h"
-#include "GroupnameContainer.h"
 #include <IMP/Restraint.h>
 #include <IMP/GroupnameScore.h>
+#include <IMP/GroupnameContainer.h>
 
+
 IMPCORE_BEGIN_NAMESPACE
 
 //! Score based on the maximum over a set of Groupnames
Index: modules/core/tools/container_templates/function.hpp
===================================================================
--- modules/core/tools/container_templates/function.hpp	(revision 1087)
+++ modules/core/tools/container_templates/function.hpp	(working copy)
@@ -7,28 +7,25 @@
  *  Copyright 2007-8 Sali Lab. All rights reserved.
  */
 
-#ifndef IMPCORE_GROUPNAME_MODIFIER_H
-#define IMPCORE_GROUPNAME_MODIFIER_H
+#ifndef IMP_GROUPNAME_MODIFIER_H
+#define IMP_GROUPNAME_MODIFIER_H
 
-#include "core_exports.h"
-#include "internal/core_version_info.h"
+#include "config.h"
+#include "internal/kernel_version_info.h"
 #include "internal/container_helpers.h"
 #include "GroupnameContainer.h"
 
-#include <IMP/base_types.h>
+#include "base_types.h"
 
 IMP_BEGIN_NAMESPACE
 // to keep swig happy
 class Particle;
-IMP_END_NAMESPACE
 
-IMPCORE_BEGIN_NAMESPACE
-
 //! A base class for modifiers of Classnames
 /** The primary function of such a class is to change
     the passed particles.
  */
-class IMPCOREEXPORT GroupnameModifier : public RefCountedObject
+class IMPEXPORT GroupnameModifier : public RefCountedObject
 {
 public:
   GroupnameModifier();
@@ -56,19 +53,19 @@
 }
 
 //! Apply a GroupnameModifier to each in the Classnames
-IMPCOREEXPORT inline void apply(GroupnameModifier* f,
+IMPEXPORT inline void apply(GroupnameModifier* f,
                                Classnames &ps) {
   apply(f, ps.begin(), ps.end());
 }
 
 //! Apply a GroupnameModifier to each in the Classnames
-IMPCOREEXPORT inline void apply(GroupnameModifier* f,
+IMPEXPORT inline void apply(GroupnameModifier* f,
                                 GroupnameContainer *ps) {
   apply(f, ps->classnames_begin(), ps->classnames_end());
 }
 
 
-IMPCORE_END_NAMESPACE
+IMP_END_NAMESPACE
 
 //! Define the functions needed for a GroupnameModifier
 #define IMP_GROUPNAME_MODIFIER(version) \
@@ -76,4 +73,4 @@
 void show(std::ostream &out= std::cout) const;\
 void apply(ClassnameArguments);
 
-#endif  /* IMPCORE_GROUPNAME_MODIFIER_H */
+#endif  /* IMP_GROUPNAME_MODIFIER_H */
Index: modules/core/tools/container_templates/filter.hpp
===================================================================
--- modules/core/tools/container_templates/filter.hpp	(revision 1087)
+++ modules/core/tools/container_templates/filter.hpp	(working copy)
@@ -11,9 +11,9 @@
 #ifndef IMPCORE_FILTERED_LIST_GROUPNAME_CONTAINER_H
 #define IMPCORE_FILTERED_LIST_GROUPNAME_CONTAINER_H
 
-#include "core_exports.h"
-#include "GroupnameContainer.h"
+#include "config.h"
 #include "internal/core_version_info.h"
+#include <IMP/GroupnameContainer.h>
 
 IMPCORE_BEGIN_NAMESPACE
 
Index: modules/core/tools/make-container
===================================================================
--- modules/core/tools/make-container	(revision 1087)
+++ modules/core/tools/make-container	(working copy)
@@ -17,7 +17,7 @@
 
 
 def make_one(name, value, multi, args):
-    path= os.path.dirname(os.path.dirname(sys.argv[0]))
+    path= os.getcwd()+"/"
     cname= name.title().replace(' ', '')
     lcvalue= value.lower().replace(' ', '_').replace('*','')
     lcname= multi.lower().replace(' ', '_')
@@ -26,48 +26,48 @@
     
   
 
-    for p in (("include/"+multi + "Container.h",
-               path+"/tools/container_templates/container.hpp"),
-              ("src/"+multi + "Container.cpp",
-               path+"/tools/container_templates/container.cc"),
-              ("include/List"+multi + "Container.h",
-               path+"/tools/container_templates/list.hpp"),
-              ("src/List"+multi + "Container.cpp",
-               path+"/tools/container_templates/list.cc"),
-              ("include/"+multi + "sRestraint.h",
-               path+"/tools/container_templates/restraint.hpp"),
-              ("src/"+multi + "sRestraint.cpp",
-               path+"/tools/container_templates/restraint.cc"),
-              ("include/FilteredList"+multi + "Container.h",
-               path+"/tools/container_templates/filter.hpp"),
-              ("src/FilteredList"+multi + "Container.cpp",
-               path+"/tools/container_templates/filter.cc"),
-              ("include/"+multi + "ContainerSet.h",
-               path+"/tools/container_templates/set.hpp"),
-              ("src/"+multi + "ContainerSet.cpp",
-               path+"/tools/container_templates/set.cc"),
-              ("include/"+multi + "sScoreState.h",
-               path+"/tools/container_templates/state.hpp"),
-              ("src/"+multi + "sScoreState.cpp",
-               path+"/tools/container_templates/state.cc"),
-              ("include/"+multi + "Modifier.h",
-               path+"/tools/container_templates/function.hpp"),
-              ("src/"+multi + "Modifier.cpp",
-               path+"/tools/container_templates/function.cc"),
-              ("include/Minimum"+multi + "ScoreRestraint.h",
-               path+"/tools/container_templates/min.hpp"),
-              ("src/Minimum"+multi + "ScoreRestraint.cpp",
-               path+"/tools/container_templates/min.cc"),
-              ("include/Maximum"+multi + "ScoreRestraint.h",
-               path+"/tools/container_templates/max.hpp"),
-              ("src/Maximum"+multi + "ScoreRestraint.cpp",
-               path+"/tools/container_templates/max.cc"),
-              ("test/containers/test_"+cname + "_restraint.py",
-               path+"/tools/container_templates/test.py"),
-              ("pyext/"+lcname+"_containers.i",
-               path+"/tools/container_templates/swig.i"),
-              ("test/containers/test_"+cname + "_state.py",
-               path+"/tools/container_templates/test_state.py"),
+    for p in (("kernel/include/"+multi + "Container.h",
+               path+"modules/core/tools/container_templates/container.hpp"),
+              ("kernel/src/"+multi + "Container.cpp",
+               path+"/modules/core/tools/container_templates/container.cc"),
+              ("modules/core/include/List"+multi + "Container.h",
+               path+"/modules/core/tools/container_templates/list.hpp"),
+              ("modules/core/src/List"+multi + "Container.cpp",
+               path+"/modules/core/tools/container_templates/list.cc"),
+              ("modules/core/include/"+multi + "sRestraint.h",
+               path+"/modules/core/tools/container_templates/restraint.hpp"),
+              ("modules/core/src/"+multi + "sRestraint.cpp",
+               path+"/modules/core/tools/container_templates/restraint.cc"),
+              ("modules/core/include/FilteredList"+multi + "Container.h",
+               path+"/modules/core/tools/container_templates/filter.hpp"),
+              ("modules/core/src/FilteredList"+multi + "Container.cpp",
+               path+"/modules/core/tools/container_templates/filter.cc"),
+              ("modules/core/include/"+multi + "ContainerSet.h",
+               path+"/modules/core/tools/container_templates/set.hpp"),
+              ("modules/core/src/"+multi + "ContainerSet.cpp",
+               path+"/modules/core/tools/container_templates/set.cc"),
+              ("modules/core/include/"+multi + "sScoreState.h",
+               path+"/modules/core/tools/container_templates/state.hpp"),
+              ("modules/core/src/"+multi + "sScoreState.cpp",
+               path+"/modules/core/tools/container_templates/state.cc"),
+              ("kernel/include/"+multi + "Modifier.h",
+               path+"/modules/core/tools/container_templates/function.hpp"),
+              ("kernel/src/"+multi + "Modifier.cpp",
+               path+"/modules/core/tools/container_templates/function.cc"),
+              ("modules/core/include/Minimum"+multi + "ScoreRestraint.h",
+               path+"/modules/core/tools/container_templates/min.hpp"),
+              ("modules/core/src/Minimum"+multi + "ScoreRestraint.cpp",
+               path+"/modules/core/tools/container_templates/min.cc"),
+              ("modules/core/include/Maximum"+multi + "ScoreRestraint.h",
+               path+"/modules/core/tools/container_templates/max.hpp"),
+              ("modules/core/src/Maximum"+multi + "ScoreRestraint.cpp",
+               path+"/modules/core/tools/container_templates/max.cc"),
+              ("modules/core/test/containers/test_"+cname + "_restraint.py",
+               path+"/modules/core/tools/container_templates/test.py"),
+              ("modules/core/pyext/"+lcname+"_containers.i",
+               path+"/modules/core/tools/container_templates/swig.i"),
+              ("modules/core/test/containers/test_"+cname + "_state.py",
+               path+"/modules/core/tools/container_templates/test_state.py"),
               ):
         ofile= open(p[0], 'w')
         ifile= open(p[1], 'r')
Index: modules/core/include/ListPairContainer.h
===================================================================
--- modules/core/include/ListPairContainer.h	(revision 1087)
+++ modules/core/include/ListPairContainer.h	(working copy)
@@ -11,8 +11,8 @@
 #define IMPCORE_LIST_PAIR_CONTAINER_H
 
 #include "config.h"
-#include "PairContainer.h"
 #include "internal/core_version_info.h"
+#include <IMP/PairContainer.h>
 
 IMPCORE_BEGIN_NAMESPACE
 
Index: modules/core/include/CentroidOfRefinedSingletonModifier.h
===================================================================
--- modules/core/include/CentroidOfRefinedSingletonModifier.h	(revision 1087)
+++ modules/core/include/CentroidOfRefinedSingletonModifier.h	(working copy)
@@ -12,11 +12,12 @@
 #include "config.h"
 #include "internal/core_version_info.h"
 
-#include "SingletonModifier.h"
 #include "XYZDecorator.h"
 #include <IMP/ParticleRefiner.h>
 #include <IMP/Pointer.h>
+#include <IMP/SingletonModifier.h>
 
+
 IMPCORE_BEGIN_NAMESPACE
 
 //! Set the coordinates of the particle to be the centoid of the particles.
Index: modules/core/include/ClosePairsFinder.h
===================================================================
--- modules/core/include/ClosePairsFinder.h	(revision 1087)
+++ modules/core/include/ClosePairsFinder.h	(working copy)
@@ -10,16 +10,15 @@
 
 #include "config.h"
 #include "XYZRDecorator.h"
-#include "SingletonContainer.h"
 #include "FilteredListPairContainer.h"
 
 #include <IMP/RefCountedObject.h>
+#include <IMP/SingletonContainer.h>
 
 IMPCORE_BEGIN_NAMESPACE
 
 // to make SWIG happy
 class FilteredListPairContainer;
-class SingletonContainer;
 
 //! A base class for algorithms to find spatial proximities
 /** In general, the algorithm should make sure it returns all
Index: modules/core/include/DerivativesFromRefinedSingletonModifier.h
===================================================================
--- modules/core/include/DerivativesFromRefinedSingletonModifier.h	(revision 1087)
+++ modules/core/include/DerivativesFromRefinedSingletonModifier.h	(working copy)
@@ -11,11 +11,12 @@
 #include "config.h"
 #include "internal/core_version_info.h"
 
-#include "SingletonModifier.h"
 #include "XYZDecorator.h"
 #include <IMP/ParticleRefiner.h>
 #include <IMP/Pointer.h>
+#include <IMP/SingletonModifier.h>
 
+
 IMPCORE_BEGIN_NAMESPACE
 
 //! Accumulate the derivatives of the refined particles.
Index: modules/core/include/PairsRestraint.h
===================================================================
--- modules/core/include/PairsRestraint.h	(revision 1087)
+++ modules/core/include/PairsRestraint.h	(working copy)
@@ -14,11 +14,11 @@
 
 #include "config.h"
 #include "internal/core_version_info.h"
-#include "PairContainer.h"
 
 #include <IMP/Restraint.h>
 #include <IMP/Pointer.h>
 #include <IMP/PairScore.h>
+#include <IMP/PairContainer.h>
 
 #include <iostream>
 
Index: modules/core/include/MinimumPairScoreRestraint.h
===================================================================
--- modules/core/include/MinimumPairScoreRestraint.h	(revision 1087)
+++ modules/core/include/MinimumPairScoreRestraint.h	(working copy)
@@ -13,9 +13,9 @@
 
 #include "config.h"
 #include "internal/core_version_info.h"
-#include "PairContainer.h"
 #include <IMP/Restraint.h>
 #include <IMP/PairScore.h>
+#include <IMP/PairContainer.h>
 
 IMPCORE_BEGIN_NAMESPACE
 
Index: modules/core/include/FilteredListSingletonContainer.h
===================================================================
--- modules/core/include/FilteredListSingletonContainer.h	(revision 1087)
+++ modules/core/include/FilteredListSingletonContainer.h	(working copy)
@@ -12,8 +12,8 @@
 #define IMPCORE_FILTERED_LIST_SINGLETON_CONTAINER_H
 
 #include "config.h"
-#include "SingletonContainer.h"
 #include "internal/core_version_info.h"
+#include <IMP/SingletonContainer.h>
 
 IMPCORE_BEGIN_NAMESPACE
 
Index: modules/core/include/MaximumPairScoreRestraint.h
===================================================================
--- modules/core/include/MaximumPairScoreRestraint.h	(revision 1087)
+++ modules/core/include/MaximumPairScoreRestraint.h	(working copy)
@@ -13,10 +13,11 @@
 
 #include "config.h"
 #include "internal/core_version_info.h"
-#include "PairContainer.h"
 #include <IMP/Restraint.h>
 #include <IMP/PairScore.h>
+#include <IMP/PairContainer.h>
 
+
 IMPCORE_BEGIN_NAMESPACE
 
 //! Score based on the maximum over a set of Pairs
Index: modules/core/include/internal/IndexingIterator.h
===================================================================
--- modules/core/include/internal/IndexingIterator.h	(revision 1087)
+++ modules/core/include/internal/IndexingIterator.h	(working copy)
@@ -1,123 +0,0 @@
-/**
- *  \file IndexingIterator.h   \brief A class to represent a voxel grid.
- *
- *  Copyright 2007-8 Sali Lab. All rights reserved.
- *
- */
-
-#ifndef IMPCORE_INDEXING_ITERATOR_H
-#define IMPCORE_INDEXING_ITERATOR_H
-
-#include "../macros.h"
-
-#include <IMP/RefCountedObject.h>
-
-#include <iterator>
-#include <limits>
-
-IMPCORE_BEGIN_INTERNAL_NAMESPACE
-
-template <class Accessor>
-class IndexingIterator {
-  Accessor a_;
-  unsigned int i_;
-  typedef IndexingIterator<Accessor> This;
-  bool is_default() const {
-    return i_!= std::numeric_limits<unsigned int>::max();
-  }
-public:
-  typedef const typename Accessor::result_type  value_type;
-  typedef unsigned int difference_type;
-  typedef const typename Accessor::result_type reference;
-  typedef const typename Accessor::result_type* pointer;
-  typedef std::random_access_iterator_tag iterator_category;
-
-  IndexingIterator(): i_(std::numeric_limits<unsigned int>::max()){}
-  IndexingIterator(Accessor a): a_(a), i_(0){}
-  IndexingIterator(Accessor a,
-                   unsigned int i): a_(a), i_(i){}
-
-  // check which is which
-  const This& operator++() {
-    ++i_;
-    return *this;
-  }
-
-  // check which is which
-  This operator++(int) {
-    This o= *this;
-    ++i_;
-    return o;
-  }
-
-  This operator+(unsigned int i) const {
-    return This(a_, i_+i);
-  }
-
-  const This& operator+=(unsigned int i) {
-    i_+= i;
-    return *this;
-  }
-
-  unsigned int operator-(const This &o) const {
-    IMP_assert(a_== o.a_, "Don't subtract iterators from different containers");
-    return i_- o.i_;
-  }
-
-  bool operator==(const This &o) const {
-    IMP_assert(a_ == o.a_,
-               "Can only compare iterators from the same container");
-    return (i_== o.i_);
-  }
-
-  bool operator!=(const This &o) const {
-    return (i_!= o.i_);
-  }
-
-  bool operator<(const This &o) const {
-    IMP_assert(a_ == o.a_,
-               "Can only compare iterators from the same container");
-    IMP_assert(!is_default() && !o.is_default(),
-               "Ordering with uninitialized index is undefined");
-    return (i_< o.i_);
-  }
-
-  bool operator>(const This &o) const {
-    IMP_assert(a_ == o.a_,
-               "Can only compare iterators from the same container");
-    IMP_assert(!is_default() && !o.is_default(),
-               "Ordering with uninitialized index is undefined");
-    return (i_> o.i_);
-  }
-
-  bool operator>=(const This &o) const {
-    IMP_assert(a_ == o.a_,
-               "Can only compare iterators from the same container");
-    IMP_assert(!is_default() && !o.is_default(),
-               "Ordering with uninitialized index is undefined");
-    return (i_>= o.i_);
-  }
-
-  bool operator<=(const This &o) const {
-    IMP_assert(a_ == o.a_,
-               "Can only compare iterators from the same container");
-    IMP_assert(!is_default() && !o.is_default(),
-               "Ordering with uninitialized index is undefined");
-    return (i_<= o.i_);
-  }
-
-  reference operator*() const {
-    return a_(i_);
-  }
-
-  pointer operator->() const {
-    static value_type vt;
-    vt= a_(i_);
-    return &vt;
-  }
-};
-
-
-IMPCORE_END_INTERNAL_NAMESPACE
-
-#endif  /* IMPCORE_INDEXING_ITERATOR_H */
Index: modules/core/include/internal/SConscript
===================================================================
--- modules/core/include/internal/SConscript	(revision 1087)
+++ modules/core/include/internal/SConscript	(working copy)
@@ -1,12 +1,10 @@
 files = [
          'ArrayOnAttributesHelper.h',
          'Grid3D.h',
-         'IndexingIterator.h',
          'MinimalSet.h',
          'ParticleGrid.h',
          'bbox_nbl_helpers.h',
          'bond_helpers.h',
-         'container_helpers.h',
          'deprecation.h',
          'evaluate_distance_pair_score.h',
          'graph_base.h',
Index: modules/core/include/internal/container_helpers.h
===================================================================
--- modules/core/include/internal/container_helpers.h	(revision 1087)
+++ modules/core/include/internal/container_helpers.h	(working copy)
@@ -1,88 +0,0 @@
-/**
- *  \file container_helpers.h
- *  \brief Internal helpers for container classes.
- *
- *  Copyright 2007-8 Sali Lab. All rights reserved.
- *
- */
-
-#ifndef IMPCORE_INTERNAL_CONTAINER_HELPERS_H
-#define IMPCORE_INTERNAL_CONTAINER_HELPERS_H
-
-#include "../macros.h"
-#include <IMP/Particle.h>
-#include <IMP/SingletonScore.h>
-#include <IMP/PairScore.h>
-
-#include <boost/tuple/tuple.hpp>
-
-IMPCORE_BEGIN_INTERNAL_NAMESPACE
-
-template <class P>
-struct ContainerTraits {
-  static const bool is_singleton=false;
-};
-
-template <>
-struct ContainerTraits<Particle> {
-  static const bool is_singleton=true;
-  static Particles create_set(Particle*p) {return Particles();}
-  static Float evaluate(SingletonScore *ss,
-                        Particle *p,
-                        DerivativeAccumulator *ac) {
-    return ss->evaluate(p, ac);
-  }
-  template <class SM>
-  static void apply(SM *ss,
-                    Particle *p) {
-    ss->apply(p);
-  }
-};
-
-template <>
-struct ContainerTraits<ParticlePair> {
-  static const bool is_singleton=false;
-  static Particles create_set(ParticlePair p) {
-    Particles ret;
-    ret.push_back(p.first);
-    ret.push_back(p.second);
-    return ret;
-  }
-  static Float evaluate(PairScore *ss,
-                        ParticlePair p,
-                        DerivativeAccumulator *ac) {
-    return ss->evaluate(p.first, p.second, ac);
-  }
-  template <class PM>
-  static void apply(PM *ss,
-                    const ParticlePair &p) {
-    ss->apply(p.first, p.second);
-  }
-};
-
-/*template <>
-struct ContainerTraits<ParticleTriplet> {
-  static const bool is_singleton=false;
-  static Particles create_set(ParticleTriplet p) {
-    Particles ret;
-    ret.push_back(p.first);
-    ret.push_back(p.second);
-    ret.push_back(p.third);
-    return ret;
-  }
-  static Float evaluate(TripletScore *ss,
-                        ParticleTriplet p,
-                        DerivativeAccumulator *ac) {
-    return ss->evaluate(p.first, p.second, p.third, ac);
-  }
-  static void apply(TripletModifier *ss,
-                     const ParticleTriplet &p) {
-    ss->apply(p.first, p.second, p.third);
-  }
-
-};*/
-
-
-IMPCORE_END_INTERNAL_NAMESPACE
-
-#endif /* IMPCORE_INTERNAL_CONTAINER_HELPERS_H */
Index: modules/core/include/AllPairsPairContainer.h
===================================================================
--- modules/core/include/AllPairsPairContainer.h	(revision 1087)
+++ modules/core/include/AllPairsPairContainer.h	(working copy)
@@ -12,9 +12,10 @@
 #define IMPCORE_ALL_PAIRS_PAIR_CONTAINER_H
 
 #include "config.h"
-#include "PairContainer.h"
-#include "SingletonContainer.h"
+
 #include "internal/core_version_info.h"
+#include <IMP/PairContainer.h>
+#include <IMP/SingletonContainer.h>
 
 IMPCORE_BEGIN_NAMESPACE
 
Index: modules/core/include/MaximumSingletonScoreRestraint.h
===================================================================
--- modules/core/include/MaximumSingletonScoreRestraint.h	(revision 1087)
+++ modules/core/include/MaximumSingletonScoreRestraint.h	(working copy)
@@ -13,10 +13,11 @@
 
 #include "config.h"
 #include "internal/core_version_info.h"
-#include "SingletonContainer.h"
 #include <IMP/Restraint.h>
 #include <IMP/SingletonScore.h>
+#include <IMP/SingletonContainer.h>
 
+
 IMPCORE_BEGIN_NAMESPACE
 
 //! Score based on the maximum over a set of Singletons
Index: modules/core/include/ClosePairsScoreState.h
===================================================================
--- modules/core/include/ClosePairsScoreState.h	(revision 1087)
+++ modules/core/include/ClosePairsScoreState.h	(working copy)
@@ -10,10 +10,11 @@
 #include "config.h"
 #include "MaximumChangeScoreState.h"
 #include "ClosePairsFinder.h"
-#include "SingletonContainer.h"
 #include "FilteredListPairContainer.h"
 #include <IMP/ScoreState.h>
 
+#include <IMP/SingletonContainer.h>
+
 IMPCORE_BEGIN_NAMESPACE
 
 // for SWIG
Index: modules/core/include/BondPairContainer.h
===================================================================
--- modules/core/include/BondPairContainer.h	(revision 1087)
+++ modules/core/include/BondPairContainer.h	(working copy)
@@ -9,10 +9,11 @@
 #define IMPCORE_BOND_PAIR_CONTAINER_H
 
 #include "config.h"
-#include "PairContainer.h"
 #include "bond_decorators.h"
 #include "internal/core_version_info.h"
 
+#include <IMP/PairContainer.h>
+
 IMPCORE_BEGIN_NAMESPACE
 
 //! A container that pretends to contain all bonds.
Index: modules/core/include/SingletonContainer.h
===================================================================
--- modules/core/include/SingletonContainer.h	(revision 1087)
+++ modules/core/include/SingletonContainer.h	(working copy)
@@ -1,101 +0,0 @@
-/**
- *  \file SingletonContainer.h    \brief A container for particles.
- *
- *  This file is generated by a script (core/tools/make-container).
- *  Do not edit directly.
- *
- *  Copyright 2007-8 Sali Lab. All rights reserved.
- */
-
-#ifndef IMPCORE_SINGLETON_CONTAINER_H
-#define IMPCORE_SINGLETON_CONTAINER_H
-
-#include "config.h"
-#include "internal/IndexingIterator.h"
-#include "internal/container_helpers.h"
-#include <IMP/Particle.h>
-#include <IMP/base_types.h>
-#include <IMP/Pointer.h>
-
-IMPCORE_BEGIN_NAMESPACE
-
-//! A shared container for particles
-/** Stores a searchable shared collection of particles.
-    \ingroup restraints
- */
-class IMPCOREEXPORT SingletonContainer : public RefCountedObject
-{
-  struct Accessor {
-    typedef Accessor This;
-    typedef Particle* result_type;
-    typedef unsigned int argument_type;
-    result_type operator()(argument_type i) const {
-      return o_->get_particle(i);
-    }
-    Accessor(SingletonContainer *pc): o_(pc){}
-    Accessor(){}
-    IMP_COMPARISONS_1(o_);
-  private:
-    bool is_default() const { return false;}
-    // This should be ref counted, but swig memory management is broken
-    SingletonContainer* o_;
-  };
-
-public:
-  SingletonContainer();
-
-  //!
-  /** \note This function may be linear. Be aware of the complexity
-      bounds of your particular container.
-   */
-  virtual bool get_contains_particle(Particle* p) const =0;
-  //! return the number of particles in the container
-  /** \note this isn't always constant time
-   */
-  virtual unsigned int get_number_of_particles() const =0;
-
-  //! get one particle
-  virtual Particle* get_particle(unsigned int i) const=0;
-
-  //! print information about the container
-  virtual void show(std::ostream &out = std::cout) const;
-
-  //! provide information about who implemeneted the container
-  virtual VersionInfo get_version_info() const=0;
-
-  //! An iterator through the contents of the container
-  typedef internal::IndexingIterator<Accessor> ParticleIterator;
-  //! begin iterating through the particles
-  ParticleIterator particles_begin() const {
-    // Since I can't make the count mutable in Object
-    return
-      ParticleIterator(Accessor(const_cast<SingletonContainer*>(this)),
-                        0);
-  }
-  //! iterate through the particles
-  ParticleIterator particles_end() const {
-    return
-      ParticleIterator(Accessor(const_cast<SingletonContainer*>(this)),
-                        get_number_of_particles());
-    }
-
-  IMP_REF_COUNTED_DESTRUCTOR(SingletonContainer)
-};
-
-IMP_OUTPUT_OPERATOR(SingletonContainer);
-
-//! A collection of containers
-typedef std::vector<SingletonContainer*> SingletonContainers;
-//! The index to use when this container is store in another object
-typedef Index<SingletonContainer> SingletonContainerIndex;
-
-IMPCORE_END_NAMESPACE
-
-#define IMP_SINGLETON_CONTAINER(version_info)                      \
-  bool get_contains_particle(Particle* p) const;                      \
-  unsigned int get_number_of_particles() const;                   \
-  Particle* get_particle(unsigned int i) const;                       \
-  void show(std::ostream &out= std::cout) const;                   \
-  IMP::VersionInfo get_version_info() const { return version_info; }
-
-#endif  /* IMPCORE_SINGLETON_CONTAINER_H */
Index: modules/core/include/CoverRefinedSingletonModifier.h
===================================================================
--- modules/core/include/CoverRefinedSingletonModifier.h	(revision 1087)
+++ modules/core/include/CoverRefinedSingletonModifier.h	(working copy)
@@ -11,12 +11,13 @@
 #include "config.h"
 #include "internal/core_version_info.h"
 
-#include "SingletonModifier.h"
-#include "SingletonContainer.h"
 #include "XYZRDecorator.h"
 #include <IMP/ParticleRefiner.h>
 #include <IMP/Pointer.h>
+#include <IMP/SingletonModifier.h>
+#include <IMP/SingletonContainer.h>
 
+
 IMPCORE_BEGIN_NAMESPACE
 
 /** \brief This class sets the position and radius of each particle to
Index: modules/core/include/PairContainer.h
===================================================================
--- modules/core/include/PairContainer.h	(revision 1087)
+++ modules/core/include/PairContainer.h	(working copy)
@@ -1,101 +0,0 @@
-/**
- *  \file PairContainer.h    \brief A container for particle_pairs.
- *
- *  This file is generated by a script (core/tools/make-container).
- *  Do not edit directly.
- *
- *  Copyright 2007-8 Sali Lab. All rights reserved.
- */
-
-#ifndef IMPCORE_PAIR_CONTAINER_H
-#define IMPCORE_PAIR_CONTAINER_H
-
-#include "config.h"
-#include "internal/IndexingIterator.h"
-#include "internal/container_helpers.h"
-#include <IMP/Particle.h>
-#include <IMP/base_types.h>
-#include <IMP/Pointer.h>
-
-IMPCORE_BEGIN_NAMESPACE
-
-//! A shared container for particle_pairs
-/** Stores a searchable shared collection of particle_pairs.
-    \ingroup restraints
- */
-class IMPCOREEXPORT PairContainer : public RefCountedObject
-{
-  struct Accessor {
-    typedef Accessor This;
-    typedef ParticlePair result_type;
-    typedef unsigned int argument_type;
-    result_type operator()(argument_type i) const {
-      return o_->get_particle_pair(i);
-    }
-    Accessor(PairContainer *pc): o_(pc){}
-    Accessor(){}
-    IMP_COMPARISONS_1(o_);
-  private:
-    bool is_default() const { return false;}
-    // This should be ref counted, but swig memory management is broken
-    PairContainer* o_;
-  };
-
-public:
-  PairContainer();
-
-  //!
-  /** \note This function may be linear. Be aware of the complexity
-      bounds of your particular container.
-   */
-  virtual bool get_contains_particle_pair(ParticlePair p) const =0;
-  //! return the number of particle_pairs in the container
-  /** \note this isn't always constant time
-   */
-  virtual unsigned int get_number_of_particle_pairs() const =0;
-
-  //! get one particle_pair
-  virtual ParticlePair get_particle_pair(unsigned int i) const=0;
-
-  //! print information about the container
-  virtual void show(std::ostream &out = std::cout) const;
-
-  //! provide information about who implemeneted the container
-  virtual VersionInfo get_version_info() const=0;
-
-  //! An iterator through the contents of the container
-  typedef internal::IndexingIterator<Accessor> ParticlePairIterator;
-  //! begin iterating through the particle_pairs
-  ParticlePairIterator particle_pairs_begin() const {
-    // Since I can't make the count mutable in Object
-    return
-      ParticlePairIterator(Accessor(const_cast<PairContainer*>(this)),
-                        0);
-  }
-  //! iterate through the particle_pairs
-  ParticlePairIterator particle_pairs_end() const {
-    return
-      ParticlePairIterator(Accessor(const_cast<PairContainer*>(this)),
-                        get_number_of_particle_pairs());
-    }
-
-  IMP_REF_COUNTED_DESTRUCTOR(PairContainer)
-};
-
-IMP_OUTPUT_OPERATOR(PairContainer);
-
-//! A collection of containers
-typedef std::vector<PairContainer*> PairContainers;
-//! The index to use when this container is store in another object
-typedef Index<PairContainer> PairContainerIndex;
-
-IMPCORE_END_NAMESPACE
-
-#define IMP_PAIR_CONTAINER(version_info)                      \
-  bool get_contains_particle_pair(ParticlePair p) const;                      \
-  unsigned int get_number_of_particle_pairs() const;                   \
-  ParticlePair get_particle_pair(unsigned int i) const;                       \
-  void show(std::ostream &out= std::cout) const;                   \
-  IMP::VersionInfo get_version_info() const { return version_info; }
-
-#endif  /* IMPCORE_PAIR_CONTAINER_H */
Index: modules/core/include/CloseBipartitePairsScoreState.h
===================================================================
--- modules/core/include/CloseBipartitePairsScoreState.h	(revision 1087)
+++ modules/core/include/CloseBipartitePairsScoreState.h	(working copy)
@@ -11,10 +11,11 @@
 #include "config.h"
 #include "MaximumChangeScoreState.h"
 #include "ClosePairsFinder.h"
-#include "SingletonContainer.h"
 #include "FilteredListPairContainer.h"
 #include <IMP/ScoreState.h>
+#include <IMP/SingletonContainer.h>
 
+
 IMPCORE_BEGIN_NAMESPACE
 
 // for SWIG
Index: modules/core/include/FilteredListPairContainer.h
===================================================================
--- modules/core/include/FilteredListPairContainer.h	(revision 1087)
+++ modules/core/include/FilteredListPairContainer.h	(working copy)
@@ -12,8 +12,8 @@
 #define IMPCORE_FILTERED_LIST_PAIR_CONTAINER_H
 
 #include "config.h"
-#include "PairContainer.h"
 #include "internal/core_version_info.h"
+#include <IMP/PairContainer.h>
 
 IMPCORE_BEGIN_NAMESPACE
 
Index: modules/core/include/PairsScoreState.h
===================================================================
--- modules/core/include/PairsScoreState.h	(revision 1087)
+++ modules/core/include/PairsScoreState.h	(working copy)
@@ -13,18 +13,18 @@
 #define IMPCORE_PAIRS_SCORE_STATE_H
 
 #include "config.h"
-#include "PairContainer.h"
-#include "PairModifier.h"
 #include "internal/core_version_info.h"
-
+#include <IMP/PairContainer.h>
+#include <IMP/PairModifier.h>
 #include <IMP/ScoreState.h>
 
-IMPCORE_BEGIN_NAMESPACE
-
+IMP_BEGIN_NAMESPACE
 // for swig
 class PairContainer;
 class PairModifier;
+IMP_END_NAMESPACE
 
+IMPCORE_BEGIN_NAMESPACE
 //! Apply a PairFunction to a PairContainer to maintain an invariant
 /** \ingroup restraint
     An example showing a how to use such a score state to maintain a cover
Index: modules/core/include/MaximumChangeScoreState.h
===================================================================
--- modules/core/include/MaximumChangeScoreState.h	(revision 1087)
+++ modules/core/include/MaximumChangeScoreState.h	(working copy)
@@ -10,11 +10,11 @@
 
 #include "config.h"
 #include "internal/core_version_info.h"
-#include "SingletonContainer.h"
 
 #include <IMP/ScoreState.h>
 #include <IMP/Index.h>
 #include <IMP/Particle.h>
+#include <IMP/SingletonContainer.h>
 
 #include <vector>
 
Index: modules/core/include/SingletonsRestraint.h
===================================================================
--- modules/core/include/SingletonsRestraint.h	(revision 1087)
+++ modules/core/include/SingletonsRestraint.h	(working copy)
@@ -14,11 +14,11 @@
 
 #include "config.h"
 #include "internal/core_version_info.h"
-#include "SingletonContainer.h"
 
 #include <IMP/Restraint.h>
 #include <IMP/Pointer.h>
 #include <IMP/SingletonScore.h>
+#include <IMP/SingletonContainer.h>
 
 #include <iostream>
 
Index: modules/core/include/MinimumSingletonScoreRestraint.h
===================================================================
--- modules/core/include/MinimumSingletonScoreRestraint.h	(revision 1087)
+++ modules/core/include/MinimumSingletonScoreRestraint.h	(working copy)
@@ -13,9 +13,9 @@
 
 #include "config.h"
 #include "internal/core_version_info.h"
-#include "SingletonContainer.h"
 #include <IMP/Restraint.h>
 #include <IMP/SingletonScore.h>
+#include <IMP/SingletonContainer.h>
 
 IMPCORE_BEGIN_NAMESPACE
 
Index: modules/core/include/VRMLLogOptimizerState.h
===================================================================
--- modules/core/include/VRMLLogOptimizerState.h	(revision 1087)
+++ modules/core/include/VRMLLogOptimizerState.h	(working copy)
@@ -11,7 +11,7 @@
 
 #include "config.h"
 #include "internal/core_version_info.h"
-#include "SingletonContainer.h"
+#include <IMP/SingletonContainer.h>
 
 #include <IMP/base_types.h>
 #include <IMP/OptimizerState.h>
Index: modules/core/include/DerivativesToRefinedSingletonModifier.h
===================================================================
--- modules/core/include/DerivativesToRefinedSingletonModifier.h	(revision 1087)
+++ modules/core/include/DerivativesToRefinedSingletonModifier.h	(working copy)
@@ -11,10 +11,10 @@
 #include "config.h"
 #include "internal/core_version_info.h"
 
-#include "SingletonModifier.h"
 #include "XYZDecorator.h"
 #include <IMP/ParticleRefiner.h>
 #include <IMP/Pointer.h>
+#include <IMP/SingletonModifier.h>
 
 IMPCORE_BEGIN_NAMESPACE
 
Index: modules/core/include/ListSingletonContainer.h
===================================================================
--- modules/core/include/ListSingletonContainer.h	(revision 1087)
+++ modules/core/include/ListSingletonContainer.h	(working copy)
@@ -11,8 +11,8 @@
 #define IMPCORE_LIST_SINGLETON_CONTAINER_H
 
 #include "config.h"
-#include "SingletonContainer.h"
 #include "internal/core_version_info.h"
+#include <IMP/SingletonContainer.h>
 
 IMPCORE_BEGIN_NAMESPACE
 
Index: modules/core/include/SConscript
===================================================================
--- modules/core/include/SConscript	(revision 1087)
+++ modules/core/include/SConscript	(working copy)
@@ -65,10 +65,8 @@
          'NormalMover.h',
          'OpenCubicSpline.h',
          'PairChainRestraint.h',
-         'PairContainer.h',
          'PairContainerSet.h',
          'PairListRestraint.h',
-         'PairModifier.h',
          'PairsRestraint.h',
          'PairsScoreState.h',
          'ParticleFunction.h',
@@ -76,10 +74,8 @@
          'ResidueDecorator.h',
          'RestraintSet.h',
          'Rotation3D.h',
-         'SingletonContainer.h',
          'SingletonContainerSet.h',
          'SingletonListRestraint.h',
-         'SingletonModifier.h',
          'SingletonsRestraint.h',
          'SingletonsScoreState.h',
          'SphereDistancePairScore.h',
@@ -94,6 +90,7 @@
          'XYZDecorator.h',
          'XYZRDecorator.h',
          'bond_decorators.h',
+         'config.h',
          'deprecation.h',
          'macros.h',
          'model_io.h',
Index: modules/core/include/PairContainerSet.h
===================================================================
--- modules/core/include/PairContainerSet.h	(revision 1087)
+++ modules/core/include/PairContainerSet.h	(working copy)
@@ -12,8 +12,8 @@
 #define IMPCORE_PAIR_CONTAINER_SET_H
 
 #include "config.h"
-#include "PairContainer.h"
 #include "internal/core_version_info.h"
+#include <IMP/PairContainer.h>
 
 IMPCORE_BEGIN_NAMESPACE
 
Index: modules/core/include/PairModifier.h
===================================================================
--- modules/core/include/PairModifier.h	(revision 1087)
+++ modules/core/include/PairModifier.h	(working copy)
@@ -1,79 +0,0 @@
-/**
- *  \file PairModifier.h    \brief A Modifier on ParticlePairs
- *
- *  This file is generated by a script (core/tools/make-container).
- *  Do not edit directly.
- *
- *  Copyright 2007-8 Sali Lab. All rights reserved.
- */
-
-#ifndef IMPCORE_PAIR_MODIFIER_H
-#define IMPCORE_PAIR_MODIFIER_H
-
-#include "config.h"
-#include "internal/core_version_info.h"
-#include "internal/container_helpers.h"
-#include "PairContainer.h"
-
-#include <IMP/base_types.h>
-
-IMP_BEGIN_NAMESPACE
-// to keep swig happy
-class Particle;
-IMP_END_NAMESPACE
-
-IMPCORE_BEGIN_NAMESPACE
-
-//! A base class for modifiers of ParticlePairs
-/** The primary function of such a class is to change
-    the passed particles.
- */
-class IMPCOREEXPORT PairModifier : public RefCountedObject
-{
-public:
-  PairModifier();
-
-  virtual ~PairModifier();
-
-  /** Apply the function to a single value*/
-  virtual void apply(Particle *a, Particle *b)=0;
-
-  /** Print out information about the function, ending in a newline.*/
-  virtual void show(std::ostream &out = std::cout) const=0;
-
-  /** return information about the authors */
-  virtual VersionInfo get_version_info() const = 0;
-};
-
-IMP_OUTPUT_OPERATOR(PairModifier)
-
-//! Apply the PairModifier to each element of the sequence
-template <class It>
-void apply(PairModifier* f, It b, It e) {
-  for (It c=b; c != e; ++c) {
-    internal::ContainerTraits<ParticlePair>::apply(f, *c);
-  }
-}
-
-//! Apply a PairModifier to each in the ParticlePairs
-IMPCOREEXPORT inline void apply(PairModifier* f,
-                               ParticlePairs &ps) {
-  apply(f, ps.begin(), ps.end());
-}
-
-//! Apply a PairModifier to each in the ParticlePairs
-IMPCOREEXPORT inline void apply(PairModifier* f,
-                                PairContainer *ps) {
-  apply(f, ps->particle_pairs_begin(), ps->particle_pairs_end());
-}
-
-
-IMPCORE_END_NAMESPACE
-
-//! Define the functions needed for a PairModifier
-#define IMP_PAIR_MODIFIER(version) \
-VersionInfo get_version_info() const {return version;}\
-void show(std::ostream &out= std::cout) const;\
-void apply(Particle *a, Particle *b);
-
-#endif  /* IMPCORE_PAIR_MODIFIER_H */
Index: modules/core/include/SingletonContainerSet.h
===================================================================
--- modules/core/include/SingletonContainerSet.h	(revision 1087)
+++ modules/core/include/SingletonContainerSet.h	(working copy)
@@ -12,8 +12,8 @@
 #define IMPCORE_SINGLETON_CONTAINER_SET_H
 
 #include "config.h"
-#include "SingletonContainer.h"
 #include "internal/core_version_info.h"
+#include <IMP/SingletonContainer.h>
 
 IMPCORE_BEGIN_NAMESPACE
 
Index: modules/core/include/SingletonModifier.h
===================================================================
--- modules/core/include/SingletonModifier.h	(revision 1087)
+++ modules/core/include/SingletonModifier.h	(working copy)
@@ -1,79 +0,0 @@
-/**
- *  \file SingletonModifier.h    \brief A Modifier on Particles
- *
- *  This file is generated by a script (core/tools/make-container).
- *  Do not edit directly.
- *
- *  Copyright 2007-8 Sali Lab. All rights reserved.
- */
-
-#ifndef IMPCORE_SINGLETON_MODIFIER_H
-#define IMPCORE_SINGLETON_MODIFIER_H
-
-#include "config.h"
-#include "internal/core_version_info.h"
-#include "internal/container_helpers.h"
-#include "SingletonContainer.h"
-
-#include <IMP/base_types.h>
-
-IMP_BEGIN_NAMESPACE
-// to keep swig happy
-class Particle;
-IMP_END_NAMESPACE
-
-IMPCORE_BEGIN_NAMESPACE
-
-//! A base class for modifiers of Particles
-/** The primary function of such a class is to change
-    the passed particles.
- */
-class IMPCOREEXPORT SingletonModifier : public RefCountedObject
-{
-public:
-  SingletonModifier();
-
-  virtual ~SingletonModifier();
-
-  /** Apply the function to a single value*/
-  virtual void apply(Particle *a)=0;
-
-  /** Print out information about the function, ending in a newline.*/
-  virtual void show(std::ostream &out = std::cout) const=0;
-
-  /** return information about the authors */
-  virtual VersionInfo get_version_info() const = 0;
-};
-
-IMP_OUTPUT_OPERATOR(SingletonModifier)
-
-//! Apply the SingletonModifier to each element of the sequence
-template <class It>
-void apply(SingletonModifier* f, It b, It e) {
-  for (It c=b; c != e; ++c) {
-    internal::ContainerTraits<Particle>::apply(f, *c);
-  }
-}
-
-//! Apply a SingletonModifier to each in the Particles
-IMPCOREEXPORT inline void apply(SingletonModifier* f,
-                               Particles &ps) {
-  apply(f, ps.begin(), ps.end());
-}
-
-//! Apply a SingletonModifier to each in the Particles
-IMPCOREEXPORT inline void apply(SingletonModifier* f,
-                                SingletonContainer *ps) {
-  apply(f, ps->particles_begin(), ps->particles_end());
-}
-
-
-IMPCORE_END_NAMESPACE
-
-//! Define the functions needed for a SingletonModifier
-#define IMP_SINGLETON_MODIFIER(version) \
-VersionInfo get_version_info() const {return version;}\
-void show(std::ostream &out= std::cout) const;\
-void apply(Particle *a);
-
-#endif  /* IMPCORE_SINGLETON_MODIFIER_H */
Index: modules/core/include/SingletonsScoreState.h
===================================================================
--- modules/core/include/SingletonsScoreState.h	(revision 1087)
+++ modules/core/include/SingletonsScoreState.h	(working copy)
@@ -13,18 +13,18 @@
 #define IMPCORE_SINGLETONS_SCORE_STATE_H
 
 #include "config.h"
-#include "SingletonContainer.h"
-#include "SingletonModifier.h"
 #include "internal/core_version_info.h"
-
+#include <IMP/SingletonContainer.h>
+#include <IMP/SingletonModifier.h>
 #include <IMP/ScoreState.h>
 
-IMPCORE_BEGIN_NAMESPACE
-
+IMP_BEGIN_NAMESPACE
 // for swig
 class SingletonContainer;
 class SingletonModifier;
+IMP_END_NAMESPACE
 
+IMPCORE_BEGIN_NAMESPACE
 //! Apply a SingletonFunction to a SingletonContainer to maintain an invariant
 /** \ingroup restraint
     An example showing a how to use such a score state to maintain a cover
Index: modules/core/src/SingletonModifier.cpp
===================================================================
--- modules/core/src/SingletonModifier.cpp	(revision 1087)
+++ modules/core/src/SingletonModifier.cpp	(working copy)
@@ -1,20 +0,0 @@
-/**
- *  \file SingletonModifier.cpp   \brief A function on Particles.
- *
- *  This file is generated by a script (core/tools/make-container).
- *  Do not edit directly.
- *
- *  Copyright 2007-8 Sali Lab. All rights reserved.
- *
- */
-
-#include "IMP/core/SingletonModifier.h"
-
-IMPCORE_BEGIN_NAMESPACE
-
-SingletonModifier::SingletonModifier(){
-}
-
-SingletonModifier::~SingletonModifier(){}
-
-IMPCORE_END_NAMESPACE
Index: modules/core/src/PairContainer.cpp
===================================================================
--- modules/core/src/PairContainer.cpp	(revision 1087)
+++ modules/core/src/PairContainer.cpp	(working copy)
@@ -1,25 +0,0 @@
-/**
- *  \file PairContainer.cpp   \brief Container for particle_pair.
- *
- *  This file is generated by a script (core/tools/make-container).
- *  Do not edit directly.
- *
- *  Copyright 2007-8 Sali Lab. All rights reserved.
- *
- */
-
-#include "IMP/core/PairContainer.h"
-
-
-IMPCORE_BEGIN_NAMESPACE
-
-PairContainer::PairContainer(){}
-
-PairContainer::~PairContainer(){}
-
-void PairContainer::show(std::ostream &out) const {
-  out << "PairContainer" << std::endl;
-}
-
-
-IMPCORE_END_NAMESPACE
Index: modules/core/src/MinimumPairScoreRestraint.cpp
===================================================================
--- modules/core/src/MinimumPairScoreRestraint.cpp	(revision 1087)
+++ modules/core/src/MinimumPairScoreRestraint.cpp	(working copy)
@@ -11,7 +11,7 @@
 
 #include "IMP/core/MinimumPairScoreRestraint.h"
 #include "IMP/core/internal/MinimalSet.h"
-#include "IMP/core/internal/container_helpers.h"
+#include <IMP/internal/container_helpers.h>
 
 
 IMPCORE_BEGIN_NAMESPACE
@@ -31,7 +31,7 @@
             << std::distance(b,e) << std::endl);
     MS bestn(n);
     for (It it= b; it != e; ++it) {
-      float score= internal::ContainerTraits<ParticlePair>
+      float score= IMP::internal::ContainerTraits<ParticlePair>
                            ::evaluate(f, *it, NULL);
 
       if (bestn.can_insert(score)) {
@@ -53,7 +53,7 @@
   float score=0;
   for (unsigned int i=0; i< bestn.size(); ++i) {
     if (da) {
-      internal::ContainerTraits<ParticlePair>::evaluate(f_.get(),
+      IMP::internal::ContainerTraits<ParticlePair>::evaluate(f_.get(),
                                                      *bestn[i].second, da);
     }
     score+= bestn[i].first;
@@ -77,7 +77,7 @@
                              c_->particle_pairs_end(), f_.get(), n_);
   ParticlesList ret;
   for (unsigned int i=0; i< bestn.size(); ++i) {
-    Particles c= internal::ContainerTraits<ParticlePair>
+    Particles c= IMP::internal::ContainerTraits<ParticlePair>
                  ::create_set(*bestn[i].second);
     ret.push_back(c);
   }
Index: modules/core/src/MaximumPairScoreRestraint.cpp
===================================================================
--- modules/core/src/MaximumPairScoreRestraint.cpp	(revision 1087)
+++ modules/core/src/MaximumPairScoreRestraint.cpp	(working copy)
@@ -11,7 +11,7 @@
 
 #include "IMP/core/MaximumPairScoreRestraint.h"
 #include "IMP/core/internal/MinimalSet.h"
-#include "IMP/core/internal/container_helpers.h"
+#include <IMP/internal/container_helpers.h>
 
 
 IMPCORE_BEGIN_NAMESPACE
@@ -32,7 +32,7 @@
     MS bestn(n);
     for (It it= b; it != e; ++it) {
       float score
-         = internal::ContainerTraits<ParticlePair>::evaluate(f, *it, NULL);
+      = IMP::internal::ContainerTraits<ParticlePair>::evaluate(f, *it, NULL);
 
       if (bestn.can_insert(score)) {
         bestn.insert(score, it);
@@ -57,7 +57,7 @@
   float score=0;
   for (unsigned int i=0; i< bestn.size(); ++i) {
     if (da) {
-      internal::ContainerTraits<ParticlePair>::evaluate(f_.get(),
+      IMP::internal::ContainerTraits<ParticlePair>::evaluate(f_.get(),
                                                      *bestn[i].second, da);
     }
     score+= bestn[i].first;
@@ -81,7 +81,7 @@
                              c_->particle_pairs_end(), f_.get(), n_);
   ParticlesList ret;
   for (unsigned int i=0; i< bestn.size(); ++i) {
-    Particles c= internal::ContainerTraits<ParticlePair>
+    Particles c= IMP::internal::ContainerTraits<ParticlePair>
                          ::create_set(*bestn[i].second);
     ret.push_back(c);
   }
Index: modules/core/src/MinimumSingletonScoreRestraint.cpp
===================================================================
--- modules/core/src/MinimumSingletonScoreRestraint.cpp	(revision 1087)
+++ modules/core/src/MinimumSingletonScoreRestraint.cpp	(working copy)
@@ -11,7 +11,7 @@
 
 #include "IMP/core/MinimumSingletonScoreRestraint.h"
 #include "IMP/core/internal/MinimalSet.h"
-#include "IMP/core/internal/container_helpers.h"
+#include <IMP/internal/container_helpers.h>
 
 
 IMPCORE_BEGIN_NAMESPACE
@@ -31,7 +31,7 @@
             << std::distance(b,e) << std::endl);
     MS bestn(n);
     for (It it= b; it != e; ++it) {
-      float score= internal::ContainerTraits<Particle>
+      float score= IMP::internal::ContainerTraits<Particle>
                            ::evaluate(f, *it, NULL);
 
       if (bestn.can_insert(score)) {
@@ -53,7 +53,7 @@
   float score=0;
   for (unsigned int i=0; i< bestn.size(); ++i) {
     if (da) {
-      internal::ContainerTraits<Particle>::evaluate(f_.get(),
+      IMP::internal::ContainerTraits<Particle>::evaluate(f_.get(),
                                                      *bestn[i].second, da);
     }
     score+= bestn[i].first;
@@ -77,7 +77,7 @@
                              c_->particles_end(), f_.get(), n_);
   ParticlesList ret;
   for (unsigned int i=0; i< bestn.size(); ++i) {
-    Particles c= internal::ContainerTraits<Particle>
+    Particles c= IMP::internal::ContainerTraits<Particle>
                  ::create_set(*bestn[i].second);
     ret.push_back(c);
   }
Index: modules/core/src/SingletonContainer.cpp
===================================================================
--- modules/core/src/SingletonContainer.cpp	(revision 1087)
+++ modules/core/src/SingletonContainer.cpp	(working copy)
@@ -1,25 +0,0 @@
-/**
- *  \file SingletonContainer.cpp   \brief Container for particle.
- *
- *  This file is generated by a script (core/tools/make-container).
- *  Do not edit directly.
- *
- *  Copyright 2007-8 Sali Lab. All rights reserved.
- *
- */
-
-#include "IMP/core/SingletonContainer.h"
-
-
-IMPCORE_BEGIN_NAMESPACE
-
-SingletonContainer::SingletonContainer(){}
-
-SingletonContainer::~SingletonContainer(){}
-
-void SingletonContainer::show(std::ostream &out) const {
-  out << "SingletonContainer" << std::endl;
-}
-
-
-IMPCORE_END_NAMESPACE
Index: modules/core/src/PairModifier.cpp
===================================================================
--- modules/core/src/PairModifier.cpp	(revision 1087)
+++ modules/core/src/PairModifier.cpp	(working copy)
@@ -1,20 +0,0 @@
-/**
- *  \file PairModifier.cpp   \brief A function on Particles.
- *
- *  This file is generated by a script (core/tools/make-container).
- *  Do not edit directly.
- *
- *  Copyright 2007-8 Sali Lab. All rights reserved.
- *
- */
-
-#include "IMP/core/PairModifier.h"
-
-IMPCORE_BEGIN_NAMESPACE
-
-PairModifier::PairModifier(){
-}
-
-PairModifier::~PairModifier(){}
-
-IMPCORE_END_NAMESPACE
Index: modules/core/src/SingletonsScoreState.cpp
===================================================================
--- modules/core/src/SingletonsScoreState.cpp	(revision 1087)
+++ modules/core/src/SingletonsScoreState.cpp	(working copy)
@@ -11,7 +11,6 @@
  */
 
 #include "IMP/core/SingletonsScoreState.h"
-#include "IMP/core/internal/container_helpers.h"
 
 IMPCORE_BEGIN_NAMESPACE
 
Index: modules/core/src/SConscript
===================================================================
--- modules/core/src/SConscript	(revision 1087)
+++ modules/core/src/SConscript	(working copy)
@@ -60,20 +60,16 @@
          'NormalMover.cpp',
          'OpenCubicSpline.cpp',
          'PairChainRestraint.cpp',
-         'PairContainer.cpp',
          'PairContainerSet.cpp',
          'PairListRestraint.cpp',
-         'PairModifier.cpp',
          'PairsRestraint.cpp',
          'PairsScoreState.cpp',
          'QuadraticClosePairsFinder.cpp',
          'ResidueDecorator.cpp',
          'RestraintSet.cpp',
          'Rotation3D.cpp',
-         'SingletonContainer.cpp',
          'SingletonContainerSet.cpp',
          'SingletonListRestraint.cpp',
-         'SingletonModifier.cpp',
          'SingletonsRestraint.cpp',
          'SingletonsScoreState.cpp',
          'SphereDistancePairScore.cpp',
Index: modules/core/src/PairsRestraint.cpp
===================================================================
--- modules/core/src/PairsRestraint.cpp	(revision 1087)
+++ modules/core/src/PairsRestraint.cpp	(working copy)
@@ -10,7 +10,7 @@
  */
 
 #include "IMP/core/PairsRestraint.h"
-#include "IMP/core/internal/container_helpers.h"
+#include <IMP/internal/container_helpers.h>
 #include "IMP/core/ListPairContainer.h"
 
 #include <IMP/PairScore.h>
@@ -50,7 +50,7 @@
   for (PairContainer::ParticlePairIterator
          it= pc_->particle_pairs_begin();
        it != pc_->particle_pairs_end(); ++it) {
-    score += internal::ContainerTraits<ParticlePair>
+    score += IMP::internal::ContainerTraits<ParticlePair>
       ::evaluate(ss_, *it, accum);
   }
 
@@ -60,12 +60,12 @@
 
 ParticlesList PairsRestraint::get_interacting_particles() const
 {
-  if (!internal::ContainerTraits<ParticlePair>::is_singleton) {
+  if (!IMP::internal::ContainerTraits<ParticlePair>::is_singleton) {
     ParticlesList ret;
     for (PairContainer::ParticlePairIterator it
            = pc_->particle_pairs_begin();
          it != pc_->particle_pairs_end(); ++it) {
-      ret.push_back(internal::ContainerTraits<ParticlePair>::create_set(*it));
+      ret.push_back(IMP::internal::ContainerTraits<ParticlePair>::create_set(*it));
     }
     return ret;
   } else {
Index: modules/core/src/PairsScoreState.cpp
===================================================================
--- modules/core/src/PairsScoreState.cpp	(revision 1087)
+++ modules/core/src/PairsScoreState.cpp	(working copy)
@@ -11,7 +11,6 @@
  */
 
 #include "IMP/core/PairsScoreState.h"
-#include "IMP/core/internal/container_helpers.h"
 
 IMPCORE_BEGIN_NAMESPACE
 
Index: modules/core/src/SingletonsRestraint.cpp
===================================================================
--- modules/core/src/SingletonsRestraint.cpp	(revision 1087)
+++ modules/core/src/SingletonsRestraint.cpp	(working copy)
@@ -10,7 +10,7 @@
  */
 
 #include "IMP/core/SingletonsRestraint.h"
-#include "IMP/core/internal/container_helpers.h"
+#include <IMP/internal/container_helpers.h>
 #include "IMP/core/ListSingletonContainer.h"
 
 #include <IMP/SingletonScore.h>
@@ -50,7 +50,7 @@
   for (SingletonContainer::ParticleIterator
          it= pc_->particles_begin();
        it != pc_->particles_end(); ++it) {
-    score += internal::ContainerTraits<Particle>
+    score += IMP::internal::ContainerTraits<Particle>
       ::evaluate(ss_, *it, accum);
   }
 
@@ -60,12 +60,12 @@
 
 ParticlesList SingletonsRestraint::get_interacting_particles() const
 {
-  if (!internal::ContainerTraits<Particle>::is_singleton) {
+  if (!IMP::internal::ContainerTraits<Particle>::is_singleton) {
     ParticlesList ret;
     for (SingletonContainer::ParticleIterator it
            = pc_->particles_begin();
          it != pc_->particles_end(); ++it) {
-      ret.push_back(internal::ContainerTraits<Particle>::create_set(*it));
+      ret.push_back(IMP::internal::ContainerTraits<Particle>::create_set(*it));
     }
     return ret;
   } else {
Index: modules/core/src/MaximumSingletonScoreRestraint.cpp
===================================================================
--- modules/core/src/MaximumSingletonScoreRestraint.cpp	(revision 1087)
+++ modules/core/src/MaximumSingletonScoreRestraint.cpp	(working copy)
@@ -11,7 +11,7 @@
 
 #include "IMP/core/MaximumSingletonScoreRestraint.h"
 #include "IMP/core/internal/MinimalSet.h"
-#include "IMP/core/internal/container_helpers.h"
+#include <IMP/internal/container_helpers.h>
 
 
 IMPCORE_BEGIN_NAMESPACE
@@ -32,7 +32,7 @@
     MS bestn(n);
     for (It it= b; it != e; ++it) {
       float score
-         = internal::ContainerTraits<Particle>::evaluate(f, *it, NULL);
+      = IMP::internal::ContainerTraits<Particle>::evaluate(f, *it, NULL);
 
       if (bestn.can_insert(score)) {
         bestn.insert(score, it);
@@ -57,7 +57,7 @@
   float score=0;
   for (unsigned int i=0; i< bestn.size(); ++i) {
     if (da) {
-      internal::ContainerTraits<Particle>::evaluate(f_.get(),
+      IMP::internal::ContainerTraits<Particle>::evaluate(f_.get(),
                                                      *bestn[i].second, da);
     }
     score+= bestn[i].first;
@@ -81,7 +81,7 @@
                              c_->particles_end(), f_.get(), n_);
   ParticlesList ret;
   for (unsigned int i=0; i< bestn.size(); ++i) {
-    Particles c= internal::ContainerTraits<Particle>
+    Particles c= IMP::internal::ContainerTraits<Particle>
                          ::create_set(*bestn[i].second);
     ret.push_back(c);
   }
Index: modules/core/pyext/core.i
===================================================================
--- modules/core/pyext/core.i	(revision 1087)
+++ modules/core/pyext/core.i	(working copy)
@@ -97,14 +97,7 @@
 }
 
 
-%feature("director") IMP::core::SingletonContainer;
-%feature("director") IMP::core::PairContainer;
 
-// for kernel
-%feature("director") IMP::core::SingletonModifier;
-%feature("director") IMP::core::PairModifier;
-
-
 /* Don't wrap internal classes */
 %ignore IMP::core::internal::ChildArrayTraits;
 
@@ -112,11 +105,6 @@
 %include "IMP/core/ClosePairsFinder.h"
 %include "IMP/core/Mover.h"
 %include "IMP/core/MoverBase.h"
-// should go to kernel
-%include "IMP/core/SingletonContainer.h"
-%include "IMP/core/PairContainer.h"
-%include "IMP/core/SingletonModifier.h"
-%include "IMP/core/PairModifier.h"
 
 %include "IMP/core/XYZDecorator.h"
 %include "IMP/core/BondedListScoreState.h"
@@ -223,11 +211,7 @@
     %template(BondedListIndex) ::IMP::Index<BondedListScoreState>;
     %template(BondDecorators) ::std::vector<BondDecorator>;
     %template(Movers) ::std::vector<Mover*>;
-    %template(SingletonContainers) ::std::vector<SingletonContainer*>;
     %template(XYZDecorators) ::std::vector<XYZDecorator>;
     %template(XYZRDecorators) ::std::vector<XYZRDecorator>;
-    %template(PairContainers) ::std::vector<PairContainer*>;
-    %template(SingletonContainerIndex) ::IMP::Index<SingletonContainer>;
-    %template(PairContainerIndex) ::IMP::Index<PairContainer>;
   }
 }
Index: modules/core/pyext/pair_containers.i
===================================================================
--- modules/core/pyext/pair_containers.i	(revision 1087)
+++ modules/core/pyext/pair_containers.i	(working copy)
@@ -8,11 +8,11 @@
     IMP_OWN_FIRST_SECOND_THIRD_CONSTRUCTOR(PairsScoreState)
     IMP_OWN_FIRST_SECOND_CONSTRUCTOR(MinimumPairScoreRestraint)
     IMP_OWN_FIRST_SECOND_CONSTRUCTOR(MaximumPairScoreRestraint)
-    IMPCORE_CONTAINER_SWIG(FilteredListPairContainer, PairContainer, pair_container)
+    IMP_CONTAINER_SWIG(FilteredListPairContainer, PairContainer, pair_container)
     IMP_SET_OBJECT(PairsRestraint, set_pair_container)
     IMP_SET_OBJECT(PairsScoreState, set_before_evaluate_modifier)
     IMP_SET_OBJECT(PairsScoreState, set_after_evaluate_modifier)
-    IMPCORE_CONTAINER_SWIG(PairContainerSet, PairContainer, pair_container)
+    IMP_CONTAINER_SWIG(PairContainerSet, PairContainer, pair_container)
 
     }
 }
Index: modules/core/pyext/singleton_containers.i
===================================================================
--- modules/core/pyext/singleton_containers.i	(revision 1087)
+++ modules/core/pyext/singleton_containers.i	(working copy)
@@ -8,11 +8,11 @@
     IMP_OWN_FIRST_SECOND_THIRD_CONSTRUCTOR(SingletonsScoreState)
     IMP_OWN_FIRST_SECOND_CONSTRUCTOR(MinimumSingletonScoreRestraint)
     IMP_OWN_FIRST_SECOND_CONSTRUCTOR(MaximumSingletonScoreRestraint)
-    IMPCORE_CONTAINER_SWIG(FilteredListSingletonContainer, SingletonContainer, singleton_container)
+    IMP_CONTAINER_SWIG(FilteredListSingletonContainer, SingletonContainer, singleton_container)
     IMP_SET_OBJECT(SingletonsRestraint, set_singleton_container)
     IMP_SET_OBJECT(SingletonsScoreState, set_before_evaluate_modifier)
     IMP_SET_OBJECT(SingletonsScoreState, set_after_evaluate_modifier)
-    IMPCORE_CONTAINER_SWIG(SingletonContainerSet, SingletonContainer, singleton_container)
+    IMP_CONTAINER_SWIG(SingletonContainerSet, SingletonContainer, singleton_container)
 
     }
 }
Index: modules/search/tools/templates/map.hpp
===================================================================
--- modules/search/tools/templates/map.hpp	(revision 1087)
+++ modules/search/tools/templates/map.hpp	(working copy)
@@ -14,7 +14,7 @@
 #include "config.h"
 #include "internal/search_version_info.h"
 #include "internal/MapHelper.h"
-#include <IMP/core/SingletonContainer.h>
+#include <IMP/SingletonContainer.h>
 #include <IMP/core/config.h>
 #include <IMP/Particle.h>
 #include <IMP/ScoreState.h>
@@ -22,11 +22,12 @@
 #include <iostream>
 #include <algorithm>
 
+// switch to using tuples rather than multiple arguments
+#include <boost/tuple/tuple.hpp>
+
 namespace IMP {
-  namespace core {
-    // for swig
-    class SingletonContainer;
-  }
+  // for swig
+  class SingletonContainer;
 }
 
 IMPSEARCH_BEGIN_NAMESPACE
@@ -62,7 +63,7 @@
   };
   typedef std::vector<Bin> Map;
   Map map_;
-  Pointer<IMP::core::SingletonContainer> pc_;
+  Pointer<IMP::SingletonContainer> pc_;
   Keys keys_;
 
   typename Map::const_iterator get_iterator(const Values &vs) const {
@@ -103,7 +104,7 @@
       must match the number of provided value types in the template
       parameter list.
   */
-  AttributeMapNUMScoreState(IMP::core::SingletonContainer* pc,
+  AttributeMapNUMScoreState(IMP::SingletonContainer* pc,
                           KEYARGS): pc_(pc) {
     keys_= Keys(KEYPARAMS);
   }
@@ -157,7 +158,7 @@
   virtual void do_before_evaluate() {
     map_.clear();
     map_.reserve(pc_->get_number_of_particles());
-    for (IMP::core::SingletonContainer::ParticleIterator
+    for (IMP::SingletonContainer::ParticleIterator
            it= pc_->particles_begin();
          it != pc_->particles_end(); ++it) {
       Values vs= get_values(*it);
@@ -165,7 +166,7 @@
     }
     std::sort(map_.begin(), map_.end());
     map_.erase(std::unique(map_.begin(), map_.end()), map_.end());
-    for (IMP::core::SingletonContainer::ParticleIterator
+    for (IMP::SingletonContainer::ParticleIterator
            it= pc_->particles_begin();
          it != pc_->particles_end(); ++it) {
       Values vs= get_values(*it);
Index: modules/search/include/IntAttributeMapScoreState.h
===================================================================
--- modules/search/include/IntAttributeMapScoreState.h	(revision 1087)
+++ modules/search/include/IntAttributeMapScoreState.h	(working copy)
@@ -28,7 +28,7 @@
      must match the number of provided value types in the template
      parameter list.
      */
-  IntAttributeMapScoreState(IMP::core::SingletonContainer* pc,
+  IntAttributeMapScoreState(IMP::SingletonContainer* pc,
                             IntKey k0);
 
   ~IntAttributeMapScoreState();
Index: modules/search/include/AttributeMap1ScoreState.h
===================================================================
--- modules/search/include/AttributeMap1ScoreState.h	(revision 1087)
+++ modules/search/include/AttributeMap1ScoreState.h	(working copy)
@@ -14,7 +14,7 @@
 #include "config.h"
 #include "internal/search_version_info.h"
 #include "internal/MapHelper.h"
-#include <IMP/core/SingletonContainer.h>
+#include <IMP/SingletonContainer.h>
 #include <IMP/core/config.h>
 #include <IMP/Particle.h>
 #include <IMP/ScoreState.h>
@@ -22,11 +22,12 @@
 #include <iostream>
 #include <algorithm>
 
+// switch to using tuples rather than multiple arguments
+#include <boost/tuple/tuple.hpp>
+
 namespace IMP {
-  namespace core {
-    // for swig
-    class SingletonContainer;
-  }
+  // for swig
+  class SingletonContainer;
 }
 
 IMPSEARCH_BEGIN_NAMESPACE
@@ -64,7 +65,7 @@
   };
   typedef std::vector<Bin> Map;
   Map map_;
-  Pointer<IMP::core::SingletonContainer> pc_;
+  Pointer<IMP::SingletonContainer> pc_;
   Keys keys_;
 
   typename Map::const_iterator get_iterator(const Values &vs) const {
@@ -106,7 +107,7 @@
       must match the number of provided value types in the template
       parameter list.
   */
-  AttributeMap1ScoreState(IMP::core::SingletonContainer* pc,
+  AttributeMap1ScoreState(IMP::SingletonContainer* pc,
                           Key0 k0): pc_(pc) {
     keys_= Keys( k0);
   }
@@ -160,7 +161,7 @@
   virtual void do_before_evaluate() {
     map_.clear();
     map_.reserve(pc_->get_number_of_particles());
-    for (IMP::core::SingletonContainer::ParticleIterator
+    for (IMP::SingletonContainer::ParticleIterator
            it= pc_->particles_begin();
          it != pc_->particles_end(); ++it) {
       Values vs= get_values(*it);
@@ -168,7 +169,7 @@
     }
     std::sort(map_.begin(), map_.end());
     map_.erase(std::unique(map_.begin(), map_.end()), map_.end());
-    for (IMP::core::SingletonContainer::ParticleIterator
+    for (IMP::SingletonContainer::ParticleIterator
            it= pc_->particles_begin();
          it != pc_->particles_end(); ++it) {
       Values vs= get_values(*it);
Index: modules/search/include/RangeSearchScoreState.h
===================================================================
--- modules/search/include/RangeSearchScoreState.h	(revision 1090)
+++ modules/search/include/RangeSearchScoreState.h	(working copy)
@@ -13,7 +13,7 @@
 #include "internal/search_version_info.h"
 
 #include <IMP/ScoreState.h>
-#include <IMP/core/SingletonContainer.h>
+#include <IMP/SingletonContainer.h>
 
 IMPSEARCH_BEGIN_NAMESPACE
 
@@ -29,12 +29,12 @@
 class IMPSEARCHEXPORT RangeSearchScoreState:
 public ScoreState
 {
-  Pointer<IMP::core::SingletonContainer> pc_;
+  Pointer<IMP::SingletonContainer> pc_;
   FloatKeys keys_;
 public:
 
   //! Create a keys.size()-D search structure
-  RangeSearchScoreState(IMP::core::SingletonContainer* pc,
+  RangeSearchScoreState(IMP::SingletonContainer* pc,
                         const FloatKeys &keys);
 
   //! Get all particles whose attributes fall in the interval
Index: modules/search/include/AttributeMap2ScoreState.h
===================================================================
--- modules/search/include/AttributeMap2ScoreState.h	(revision 1087)
+++ modules/search/include/AttributeMap2ScoreState.h	(working copy)
@@ -14,7 +14,7 @@
 #include "config.h"
 #include "internal/search_version_info.h"
 #include "internal/MapHelper.h"
-#include <IMP/core/SingletonContainer.h>
+#include <IMP/SingletonContainer.h>
 #include <IMP/core/config.h>
 #include <IMP/Particle.h>
 #include <IMP/ScoreState.h>
@@ -22,11 +22,12 @@
 #include <iostream>
 #include <algorithm>
 
+// switch to using tuples rather than multiple arguments
+#include <boost/tuple/tuple.hpp>
+
 namespace IMP {
-  namespace core {
-    // for swig
-    class SingletonContainer;
-  }
+  // for swig
+  class SingletonContainer;
 }
 
 IMPSEARCH_BEGIN_NAMESPACE
@@ -70,7 +71,7 @@
   };
   typedef std::vector<Bin> Map;
   Map map_;
-  Pointer<IMP::core::SingletonContainer> pc_;
+  Pointer<IMP::SingletonContainer> pc_;
   Keys keys_;
 
   typename Map::const_iterator get_iterator(const Values &vs) const {
@@ -115,7 +116,7 @@
       must match the number of provided value types in the template
       parameter list.
   */
-  AttributeMap2ScoreState(IMP::core::SingletonContainer* pc,
+  AttributeMap2ScoreState(IMP::SingletonContainer* pc,
                           Key0 k0,
                           Key1 k1): pc_(pc) {
     keys_= Keys( k0,  k1);
@@ -174,7 +175,7 @@
   virtual void do_before_evaluate() {
     map_.clear();
     map_.reserve(pc_->get_number_of_particles());
-    for (IMP::core::SingletonContainer::ParticleIterator
+    for (IMP::SingletonContainer::ParticleIterator
            it= pc_->particles_begin();
          it != pc_->particles_end(); ++it) {
       Values vs= get_values(*it);
@@ -182,7 +183,7 @@
     }
     std::sort(map_.begin(), map_.end());
     map_.erase(std::unique(map_.begin(), map_.end()), map_.end());
-    for (IMP::core::SingletonContainer::ParticleIterator
+    for (IMP::SingletonContainer::ParticleIterator
            it= pc_->particles_begin();
          it != pc_->particles_end(); ++it) {
       Values vs= get_values(*it);
Index: modules/search/include/AttributeMap3ScoreState.h
===================================================================
--- modules/search/include/AttributeMap3ScoreState.h	(revision 1087)
+++ modules/search/include/AttributeMap3ScoreState.h	(working copy)
@@ -14,7 +14,7 @@
 #include "config.h"
 #include "internal/search_version_info.h"
 #include "internal/MapHelper.h"
-#include <IMP/core/SingletonContainer.h>
+#include <IMP/SingletonContainer.h>
 #include <IMP/core/config.h>
 #include <IMP/Particle.h>
 #include <IMP/ScoreState.h>
@@ -22,11 +22,12 @@
 #include <iostream>
 #include <algorithm>
 
+// switch to using tuples rather than multiple arguments
+#include <boost/tuple/tuple.hpp>
+
 namespace IMP {
-  namespace core {
-    // for swig
-    class SingletonContainer;
-  }
+  // for swig
+  class SingletonContainer;
 }
 
 IMPSEARCH_BEGIN_NAMESPACE
@@ -76,7 +77,7 @@
   };
   typedef std::vector<Bin> Map;
   Map map_;
-  Pointer<IMP::core::SingletonContainer> pc_;
+  Pointer<IMP::SingletonContainer> pc_;
   Keys keys_;
 
   typename Map::const_iterator get_iterator(const Values &vs) const {
@@ -124,7 +125,7 @@
       must match the number of provided value types in the template
       parameter list.
   */
-  AttributeMap3ScoreState(IMP::core::SingletonContainer* pc,
+  AttributeMap3ScoreState(IMP::SingletonContainer* pc,
                           Key0 k0,
                           Key1 k1,
                           Key2 k2): pc_(pc) {
@@ -188,7 +189,7 @@
   virtual void do_before_evaluate() {
     map_.clear();
     map_.reserve(pc_->get_number_of_particles());
-    for (IMP::core::SingletonContainer::ParticleIterator
+    for (IMP::SingletonContainer::ParticleIterator
            it= pc_->particles_begin();
          it != pc_->particles_end(); ++it) {
       Values vs= get_values(*it);
@@ -196,7 +197,7 @@
     }
     std::sort(map_.begin(), map_.end());
     map_.erase(std::unique(map_.begin(), map_.end()), map_.end());
-    for (IMP::core::SingletonContainer::ParticleIterator
+    for (IMP::SingletonContainer::ParticleIterator
            it= pc_->particles_begin();
          it != pc_->particles_end(); ++it) {
       Values vs= get_values(*it);
Index: modules/search/include/AttributeMap4ScoreState.h
===================================================================
--- modules/search/include/AttributeMap4ScoreState.h	(revision 1087)
+++ modules/search/include/AttributeMap4ScoreState.h	(working copy)
@@ -14,7 +14,7 @@
 #include "config.h"
 #include "internal/search_version_info.h"
 #include "internal/MapHelper.h"
-#include <IMP/core/SingletonContainer.h>
+#include <IMP/SingletonContainer.h>
 #include <IMP/core/config.h>
 #include <IMP/Particle.h>
 #include <IMP/ScoreState.h>
@@ -22,11 +22,12 @@
 #include <iostream>
 #include <algorithm>
 
+// switch to using tuples rather than multiple arguments
+#include <boost/tuple/tuple.hpp>
+
 namespace IMP {
-  namespace core {
-    // for swig
-    class SingletonContainer;
-  }
+  // for swig
+  class SingletonContainer;
 }
 
 IMPSEARCH_BEGIN_NAMESPACE
@@ -82,7 +83,7 @@
   };
   typedef std::vector<Bin> Map;
   Map map_;
-  Pointer<IMP::core::SingletonContainer> pc_;
+  Pointer<IMP::SingletonContainer> pc_;
   Keys keys_;
 
   typename Map::const_iterator get_iterator(const Values &vs) const {
@@ -133,7 +134,7 @@
       must match the number of provided value types in the template
       parameter list.
   */
-  AttributeMap4ScoreState(IMP::core::SingletonContainer* pc,
+  AttributeMap4ScoreState(IMP::SingletonContainer* pc,
                           Key0 k0,
                           Key1 k1,
                           Key2 k2,
@@ -202,7 +203,7 @@
   virtual void do_before_evaluate() {
     map_.clear();
     map_.reserve(pc_->get_number_of_particles());
-    for (IMP::core::SingletonContainer::ParticleIterator
+    for (IMP::SingletonContainer::ParticleIterator
            it= pc_->particles_begin();
          it != pc_->particles_end(); ++it) {
       Values vs= get_values(*it);
@@ -210,7 +211,7 @@
     }
     std::sort(map_.begin(), map_.end());
     map_.erase(std::unique(map_.begin(), map_.end()), map_.end());
-    for (IMP::core::SingletonContainer::ParticleIterator
+    for (IMP::SingletonContainer::ParticleIterator
            it= pc_->particles_begin();
          it != pc_->particles_end(); ++it) {
       Values vs= get_values(*it);
Index: modules/search/src/IntAttributeMapScoreState.cpp
===================================================================
--- modules/search/src/IntAttributeMapScoreState.cpp	(revision 1087)
+++ modules/search/src/IntAttributeMapScoreState.cpp	(working copy)
@@ -10,7 +10,7 @@
 IMPSEARCH_BEGIN_NAMESPACE
 
 IntAttributeMapScoreState
-::IntAttributeMapScoreState(IMP::core::SingletonContainer *pc,
+::IntAttributeMapScoreState(IMP::SingletonContainer *pc,
                             IntKey k): AttributeMap1ScoreState<Int>(pc, k){}
 
 IntAttributeMapScoreState::~IntAttributeMapScoreState(){}
Index: modules/search/SConscript
===================================================================
--- modules/search/SConscript	(revision 1087)
+++ modules/search/SConscript	(working copy)
@@ -1,2 +1,11 @@
 Import('env')
 Export('env')
+env.Append(LIBS=['imp_core'])
+env.validate()
+
+# Subdirectories:
+SConscript('src/SConscript')
+SConscript('include/SConscript')
+SConscript('pyext/SConscript')
+SConscript('test/SConscript')
+SConscript('bin/SConscript')
Index: kernel/test/run-all-tests.py
===================================================================
--- kernel/test/run-all-tests.py	(revision 1087)
+++ kernel/test/run-all-tests.py	(working copy)
@@ -1 +1,26 @@
-link ../../tools/run-all-tests.py
\ No newline at end of file
+import unittest, sys, os, re
+
+def regressionTest():
+    """Run all tests in files called test_*.py in current directory and
+       subdirectories"""
+    path = os.path.dirname(sys.argv[0])
+    if path == '':
+        path = '.'
+    # Tell test cases where to find any input files needed
+    os.environ['TEST_DIRECTORY'] = path
+    modobjs = []
+    for subdir in [''] + [x for x in os.listdir(path) \
+                          if os.path.isdir(os.path.join(path, x))]:
+        files = os.listdir(os.path.join(path, subdir))
+        test = re.compile("^test_.*\.py$", re.IGNORECASE)
+        files = filter(test.search, files)
+        modnames = [os.path.splitext(f)[0] for f in files]
+        sys.path.insert(0, os.path.join(path, subdir))
+        modobjs.extend([__import__(m) for m in modnames])
+        sys.path.pop(0)
+
+    tests = [unittest.defaultTestLoader.loadTestsFromModule(o) for o in modobjs]
+    return unittest.TestSuite(tests)
+
+if __name__ == "__main__":
+    unittest.main(defaultTest="regressionTest")
Index: kernel/test/particles/test_refcount.py
===================================================================
--- kernel/test/particles/test_refcount.py	(revision 1087)
+++ kernel/test/particles/test_refcount.py	(working copy)
@@ -34,8 +34,7 @@
         self._check_number(0)
 
         m= IMP.Model()
-        p= IMP.Particle()
-        m.add_particle(p)
+        p= IMP.Particle(m)
         self._check_number(1)
 
         p=1
@@ -47,8 +46,8 @@
         """Check that ref counting works with removing particles"""
         self._check_number(0)
         m= IMP.Model()
-        p= IMP.Particle()
-        pi= m.add_particle(p)
+        p= IMP.Particle(m)
+        pi= p.get_index()
         self._check_number(1)
         m.remove_particle(pi)
         self._check_number(1)
@@ -86,10 +85,10 @@
         self._check_number(0)
         m= IMP.Model()
         print 7
-        p= IMP.Particle()
+        p= IMP.Particle(m)
         print p
         print 8
-        pi= m.add_particle(p)
+        pi= p.get_index()
         print 9
         p=None
         self._check_number(1)
@@ -141,8 +140,8 @@
         """Check that removed particles are skipped"""
         print "skipped"
         m= IMP.Model()
-        p= IMP.Particle()
-        pi= m.add_particle(p)
+        p= IMP.Particle(m)
+        pi= p.get_index()
         ps= m.get_particles()
         self.assertEqual(len(ps), 1, "Should only be 1 particle")
         m.remove_particle(pi)
Index: kernel/include/Model.h
===================================================================
--- kernel/include/Model.h	(revision 1087)
+++ kernel/include/Model.h	(working copy)
@@ -28,6 +28,10 @@
 //! Class for storing model, its restraints, and particles.
 /** The Model maintains a standard IMP container for each of Particle,
     ScoreState and Restraint object types.
+
+    \note Think carefully about using the iterators over the entire set
+	of Particles or Restraints. Most operations should be done using
+	a user-passed set of Particles or Restraints instead.
  */
 class IMPEXPORT Model: public Object
 {
@@ -35,9 +39,10 @@
   friend class Restraint;
   unsigned int iteration_;
 public:
-  //!
+  /** Construct and empty model */
   Model();
-  //!
+  /** Particles, ScoreStates and Restraints are ref-counted so they
+      will be deleted if no other Pointers to them held. */
   ~Model();
 
   IMP_CONTAINER(Particle, particle, ParticleIndex);
Index: kernel/include/internal/IndexingIterator.h
===================================================================
--- kernel/include/internal/IndexingIterator.h	(revision 1087)
+++ kernel/include/internal/IndexingIterator.h	(working copy)
@@ -5,8 +5,8 @@
  *
  */
 
-#ifndef IMPCORE_INDEXING_ITERATOR_H
-#define IMPCORE_INDEXING_ITERATOR_H
+#ifndef IMP_INDEXING_ITERATOR_H
+#define IMP_INDEXING_ITERATOR_H
 
 #include "../macros.h"
 
@@ -15,7 +15,7 @@
 #include <iterator>
 #include <limits>
 
-IMPCORE_BEGIN_INTERNAL_NAMESPACE
+IMP_BEGIN_INTERNAL_NAMESPACE
 
 template <class Accessor>
 class IndexingIterator {
@@ -118,6 +118,6 @@
 };
 
 
-IMPCORE_END_INTERNAL_NAMESPACE
+IMP_END_INTERNAL_NAMESPACE
 
-#endif  /* IMPCORE_INDEXING_ITERATOR_H */
+#endif  /* IMP_INDEXING_ITERATOR_H */

Property changes on: kernel/include/internal/IndexingIterator.h
___________________________________________________________________
Added: svn:mergeinfo

Index: kernel/include/internal/SConscript
===================================================================
--- kernel/include/internal/SConscript	(revision 1087)
+++ kernel/include/internal/SConscript	(working copy)
@@ -1,7 +1,8 @@
 files = ['AttributeTable.h', 'Vector.h',
          'ref_counting.h', 'ObjectContainer.h',
          'kernel_version_info.h', 'constants.h', 'units.h',
-         'utility.h', 'Unit.h', 'ExponentialNumber.h', 'key_helpers.h']
+         'utility.h', 'Unit.h', 'ExponentialNumber.h', 'key_helpers.h',
+         'IndexingIterator.h', 'container_helpers.h']
 
 files = [File(f) for f in files]
 Return('files')
Index: kernel/include/internal/container_helpers.h
===================================================================
--- kernel/include/internal/container_helpers.h	(revision 1087)
+++ kernel/include/internal/container_helpers.h	(working copy)
@@ -6,17 +6,17 @@
  *
  */
 
-#ifndef IMPCORE_INTERNAL_CONTAINER_HELPERS_H
-#define IMPCORE_INTERNAL_CONTAINER_HELPERS_H
+#ifndef IMP_INTERNAL_CONTAINER_HELPERS_H
+#define IMP_INTERNAL_CONTAINER_HELPERS_H
 
-#include "../macros.h"
-#include <IMP/Particle.h>
-#include <IMP/SingletonScore.h>
-#include <IMP/PairScore.h>
+#include "../base_types.h"
+#include "../Particle.h"
+#include "../SingletonScore.h"
+#include "../PairScore.h"
 
 #include <boost/tuple/tuple.hpp>
 
-IMPCORE_BEGIN_INTERNAL_NAMESPACE
+IMP_BEGIN_INTERNAL_NAMESPACE
 
 template <class P>
 struct ContainerTraits {
@@ -83,6 +83,6 @@
 };*/
 
 
-IMPCORE_END_INTERNAL_NAMESPACE
+IMP_END_INTERNAL_NAMESPACE
 
-#endif /* IMPCORE_INTERNAL_CONTAINER_HELPERS_H */
+#endif /* IMP_INTERNAL_CONTAINER_HELPERS_H */

Property changes on: kernel/include/internal/container_helpers.h
___________________________________________________________________
Added: svn:mergeinfo

Index: kernel/include/SConscript
===================================================================
--- kernel/include/SConscript	(revision 1087)
+++ kernel/include/SConscript	(working copy)
@@ -8,7 +8,8 @@
          'UnaryFunction.h', 'PairScore.h', 'SingletonScore.h', 'macros.h',
          'TripletScore.h', 'exception.h', 'VersionInfo.h',
          'Object.h', 'Pointer.h', 'RefCountedObject.h', 'ParticleRefiner.h',
-         'container_macros.h']
+         'container_macros.h', 'SingletonModifier.h', 'PairModifier.h',
+         'SingletonContainer.h', 'PairContainer.h']
 
 # Subdirectories
 files = [File(f) for f in files] \
Index: kernel/include/Particle.h
===================================================================
--- kernel/include/Particle.h	(revision 1087)
+++ kernel/include/Particle.h	(working copy)
@@ -96,8 +96,9 @@
 
 public:
 
-
-
+  //! Construct a particle and add it to the Model
+  Particle(Model *m);
+  
   Particle();
   ~Particle();
 
Index: kernel/src/Particle.cpp
===================================================================
--- kernel/src/Particle.cpp	(revision 1087)
+++ kernel/src/Particle.cpp	(working copy)
@@ -11,6 +11,13 @@
 
 IMP_BEGIN_NAMESPACE
 
+Particle::Particle(Model *m)
+{
+  is_active_ = true;
+  m->add_particle(this);
+}
+
+
 Particle::Particle()
 {
   is_active_ = true;
Index: kernel/src/SConscript
===================================================================
--- kernel/src/SConscript	(revision 1087)
+++ kernel/src/SConscript	(working copy)
@@ -13,7 +13,9 @@
          'OptimizerState.cpp', 'Log.cpp', 'Restraint.cpp', 'Optimizer.cpp',
          'random.cpp', 'Key.cpp', 'exception.cpp', 'ParticleRefiner.cpp',
          'UnaryFunction.cpp', 'PairScore.cpp',
-         'SingletonScore.cpp', 'TripletScore.cpp'
+         'SingletonScore.cpp', 'TripletScore.cpp',
+         'SingletonModifier.cpp', 'PairModifier.cpp',
+         'SingletonContainer.cpp', 'PairContainer.cpp'
         ] + internal_files
 
 # Build the shared library:
Index: kernel/pyext/IMP/test.py
===================================================================
--- kernel/pyext/IMP/test.py	(revision 1087)
+++ kernel/pyext/IMP/test.py	(working copy)
@@ -34,8 +34,7 @@
     def create_point_particle(self, model, x, y, z):
         """Make a particle with optimizable x, y and z attributes, and
            add it to the model."""
-        p = IMP.Particle()
-        model.add_particle(p)
+        p = IMP.Particle(model)
         p.add_attribute(IMP.FloatKey("x"), x, True)
         p.add_attribute(IMP.FloatKey("y"), y, True)
         p.add_attribute(IMP.FloatKey("z"), z, True)
Index: kernel/pyext/IMP.i
===================================================================
--- kernel/pyext/IMP.i	(revision 1087)
+++ kernel/pyext/IMP.i	(working copy)
@@ -119,6 +119,11 @@
 %feature("director") IMP::TripletScore;
 %feature("director") IMP::Optimizer;
 %feature("director") IMP::ParticleRefiner;
+%feature("director") IMP::SingletonContainer;
+%feature("director") IMP::PairContainer;
+%feature("director") IMP::SingletonModifier;
+%feature("director") IMP::PairModifier;
+    
 
 %include "IMP/base_types.h"
 %include "Vector3D.i"
@@ -140,6 +145,11 @@
 %include "Particle.i"
 %include "IMP/DecoratorBase.h"
 %include "IMP/Optimizer.h"
+%include "IMP/SingletonContainer.h"
+%include "IMP/PairContainer.h"
+%include "IMP/SingletonModifier.h"
+%include "IMP/PairModifier.h"
+    
 
 namespace IMP {
   %template(ParticleIndex) Index<ParticleTag>;
@@ -161,6 +171,10 @@
   %template(Floats) ::std::vector<Float>;
   %template(Strings) ::std::vector<String>;
   %template(Ints) ::std::vector<Int>;
+  %template(SingletonContainers) ::std::vector<SingletonContainer*>;
+  %template(PairContainers) ::std::vector<PairContainer*>;
+  %template(SingletonContainerIndex) ::IMP::Index<SingletonContainer>;
+  %template(PairContainerIndex) ::IMP::Index<PairContainer>;
   /*%template(Vector3D) VectorD<3>;
   %template(random_vector_on_sphere) random_vector_on_sphere<3>;
   %template(random_vector_in_sphere) random_vector_in_sphere<3>;
