diff --git a/tools/nightly-tests/test-install/test.py b/tools/nightly-tests/test-install/test.py
index ad2ac77bc0..79725c03a8 100644
--- a/tools/nightly-tests/test-install/test.py
+++ b/tools/nightly-tests/test-install/test.py
@@ -70,7 +70,7 @@ class IMPInstallTests(unittest.TestCase):
 
     def test_applications_installed(self):
         """Check install of basic applications"""
-        apps = [App('cnmultifit', python=True), App('foxs'),
+        apps = [App('foxs'),
                 App('ligand_score'), App('multifit', python=True),
                 App('pdb_check'), App('pdb_rmf'),
                 App('rmf_cat', can_exit_1=True),
diff --git a/tools/nightly-tests/test-install/test_mock.py b/tools/nightly-tests/test-install/test_mock.py
index 82162c8f18..83f9102f34 100644
--- a/tools/nightly-tests/test-install/test_mock.py
+++ b/tools/nightly-tests/test-install/test_mock.py
@@ -35,7 +35,6 @@ class IMPMockTests(unittest.TestCase):
         _ = IMP.npctransport.Configuration
         # Check that most other modules (particularly those with many
         # dependencies) are present
-        import IMP.cnmultifit
         import IMP.domino
         import IMP.em
         import IMP.gsl
@@ -63,7 +62,7 @@ class IMPMockTests(unittest.TestCase):
                       'interface_cross_links', 'em3d_score',
                       'em3d_single_score', 'saxs_score', 'interface_rtc',
                       'nmr_rtc_score', 'soap_score']
-        apps = ['cluster_profiles', 'cnmultifit',
+        apps = ['cluster_profiles',
                 'complex_to_anchor_graph', 'compute_chi',
                 'estimate_threshold_from_molecular_mass', 'foxs',
                 'ligand_score', 'map2pca', 'mol2pca',
